// @aztec/bb.js/dest/node-cjs/bb_backends/node/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAsyncBackend = createAsyncBackend;
exports.createSyncBackend = createSyncBackend;
const native_socket_js_1 = require("./native_socket.js");
const wasm_js_1 = require("../wasm.js");
const native_shm_js_1 = require("./native_shm.js");
const sync_to_async_adapter_js_1 = require("../sync_to_async_adapter.js");
const platform_js_1 = require("./platform.js");
const index_js_1 = require("../../barretenberg/index.js");
const index_js_2 = require("../index.js");
/**
 * Create backend of specific type (no fallback)
 */
async function createAsyncBackend(type, options, logger) {
    switch (type) {
        case index_js_2.BackendType.NativeUnixSocket: {
            const bbPath = (0, platform_js_1.findBbBinary)(options.bbPath);
            if (!bbPath) {
                throw new Error('Native backend requires bb binary.');
            }
            logger(`Using native Unix socket backend: ${bbPath}`);
            const socket = new native_socket_js_1.BarretenbergNativeSocketAsyncBackend(bbPath, options.threads);
            return new index_js_1.Barretenberg(socket, options);
        }
        case index_js_2.BackendType.NativeSharedMemory: {
            const bbPath = (0, platform_js_1.findBbBinary)(options.bbPath);
            if (!bbPath) {
                throw new Error('Native backend requires bb binary.');
            }
            const napiPath = (0, platform_js_1.findNapiBinary)();
            if (!napiPath) {
                throw new Error('Native sync backend requires napi client stub.');
            }
            logger(`Using native shared memory backend (via sync adapter): ${bbPath}`);
            // Use sync backend with adapter to provide async interface
            const syncBackend = await native_shm_js_1.BarretenbergNativeShmSyncBackend.new(bbPath, options.threads, options.maxClients);
            const asyncBackend = new sync_to_async_adapter_js_1.SyncToAsyncAdapter(syncBackend);
            return new index_js_1.Barretenberg(asyncBackend, options);
        }
        case index_js_2.BackendType.Wasm:
        case index_js_2.BackendType.WasmWorker: {
            const useWorker = type === index_js_2.BackendType.WasmWorker;
            logger(`Using WASM backend (worker: ${useWorker})`);
            const wasm = await wasm_js_1.BarretenbergWasmAsyncBackend.new({
                threads: options.threads,
                wasmPath: options.wasmPath,
                logger: options.logger,
                memory: options.memory,
                useWorker,
            });
            return new index_js_1.Barretenberg(wasm, options);
        }
        default:
            throw new Error(`Unknown backend type: ${type}`);
    }
}
/**
 * Create backend of specific type (no fallback)
 */
async function createSyncBackend(type, options, logger) {
    switch (type) {
        case index_js_2.BackendType.NativeSharedMemory: {
            const bbPath = (0, platform_js_1.findBbBinary)(options.bbPath);
            if (!bbPath) {
                throw new Error('Native backend requires bb binary.');
            }
            const napiPath = (0, platform_js_1.findNapiBinary)();
            if (!napiPath) {
                throw new Error('Native sync backend requires napi client stub.');
            }
            logger(`Using native shared memory backend: ${bbPath}`);
            const shm = await native_shm_js_1.BarretenbergNativeShmSyncBackend.new(bbPath, options.threads, options.maxClients);
            return new index_js_1.BarretenbergSync(shm);
        }
        case index_js_2.BackendType.Wasm:
            logger('Using WASM backend');
            const wasm = await wasm_js_1.BarretenbergWasmSyncBackend.new(options.wasmPath, logger);
            return new index_js_1.BarretenbergSync(wasm);
        default:
            throw new Error(`Backend ${type} not supported for BarretenbergSync`);
    }
}
