// @aztec/bb.js/dest/node/cbind/generated/async.js
// AUTOGENERATED FILE - DO NOT EDIT
import { Decoder, Encoder } from 'msgpackr';
import { fromAesDecrypt, fromAesEncrypt, fromBlake2s, fromBlake2sToField, fromBn254FrSqrt, fromCircuitComputeVk, fromCircuitProve, fromCircuitStats, fromCircuitVerify, fromCircuitWriteSolidityVerifier, fromClientIvcAccumulate, fromClientIvcCheckPrecomputedVk, fromClientIvcComputeIvcVk, fromClientIvcComputeStandaloneVk, fromClientIvcLoad, fromClientIvcProve, fromClientIvcStart, fromClientIvcStats, fromClientIvcVerify, fromEcdsaSecp256k1ComputePublicKey, fromEcdsaSecp256k1ConstructSignature, fromEcdsaSecp256k1RecoverPublicKey, fromEcdsaSecp256k1VerifySignature, fromEcdsaSecp256r1ComputePublicKey, fromEcdsaSecp256r1ConstructSignature, fromEcdsaSecp256r1RecoverPublicKey, fromEcdsaSecp256r1VerifySignature, fromGrumpkinAdd, fromGrumpkinBatchMul, fromGrumpkinGetRandomFr, fromGrumpkinMul, fromGrumpkinReduce512, fromMegaVkAsFields, fromPedersenCommit, fromPedersenHash, fromPedersenHashBuffer, fromPoseidon2Hash, fromPoseidon2HashAccumulate, fromPoseidon2Permutation, fromSchnorrComputePublicKey, fromSchnorrConstructSignature, fromSchnorrVerifySignature, fromSecp256k1GetRandomFr, fromSecp256k1Mul, fromSecp256k1Reduce512, fromShutdown, fromSrsInitGrumpkinSrs, fromSrsInitSrs, fromVkAsFields, toAesDecryptResponse, toAesEncryptResponse, toBlake2sResponse, toBlake2sToFieldResponse, toBn254FrSqrtResponse, toCircuitComputeVkResponse, toCircuitInfoResponse, toCircuitProveResponse, toCircuitVerifyResponse, toCircuitWriteSolidityVerifierResponse, toClientIvcAccumulateResponse, toClientIvcCheckPrecomputedVkResponse, toClientIvcComputeIvcVkResponse, toClientIvcComputeStandaloneVkResponse, toClientIvcLoadResponse, toClientIvcProveResponse, toClientIvcStartResponse, toClientIvcStatsResponse, toClientIvcVerifyResponse, toEcdsaSecp256k1ComputePublicKeyResponse, toEcdsaSecp256k1ConstructSignatureResponse, toEcdsaSecp256k1RecoverPublicKeyResponse, toEcdsaSecp256k1VerifySignatureResponse, toEcdsaSecp256r1ComputePublicKeyResponse, toEcdsaSecp256r1ConstructSignatureResponse, toEcdsaSecp256r1RecoverPublicKeyResponse, toEcdsaSecp256r1VerifySignatureResponse, toGrumpkinAddResponse, toGrumpkinBatchMulResponse, toGrumpkinGetRandomFrResponse, toGrumpkinMulResponse, toGrumpkinReduce512Response, toMegaVkAsFieldsResponse, toPedersenCommitResponse, toPedersenHashBufferResponse, toPedersenHashResponse, toPoseidon2HashAccumulateResponse, toPoseidon2HashResponse, toPoseidon2PermutationResponse, toSchnorrComputePublicKeyResponse, toSchnorrConstructSignatureResponse, toSchnorrVerifySignatureResponse, toSecp256k1GetRandomFrResponse, toSecp256k1MulResponse, toSecp256k1Reduce512Response, toShutdownResponse, toSrsInitGrumpkinSrsResponse, toSrsInitSrsResponse, toVkAsFieldsResponse } from './api_types.js';
async function msgpackCall(backend, input) { const inputBuffer = new Encoder({ useRecords: false }).pack(input); const encodedResult = await backend.call(inputBuffer); return new Decoder({ useRecords: false }).unpack(encodedResult); }
export class AsyncApi {
    constructor(backend) {
        this.backend = backend;
    }
    circuitProve(command) {
        const msgpackCommand = fromCircuitProve(command);
        return msgpackCall(this.backend, [["CircuitProve", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'CircuitProveResponse') {
                throw new Error(`Expected variant name 'CircuitProveResponse' but got '${variantName}'`);
            }
            return toCircuitProveResponse(result);
        });
    }
    circuitComputeVk(command) {
        const msgpackCommand = fromCircuitComputeVk(command);
        return msgpackCall(this.backend, [["CircuitComputeVk", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'CircuitComputeVkResponse') {
                throw new Error(`Expected variant name 'CircuitComputeVkResponse' but got '${variantName}'`);
            }
            return toCircuitComputeVkResponse(result);
        });
    }
    circuitStats(command) {
        const msgpackCommand = fromCircuitStats(command);
        return msgpackCall(this.backend, [["CircuitStats", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'CircuitInfoResponse') {
                throw new Error(`Expected variant name 'CircuitInfoResponse' but got '${variantName}'`);
            }
            return toCircuitInfoResponse(result);
        });
    }
    circuitVerify(command) {
        const msgpackCommand = fromCircuitVerify(command);
        return msgpackCall(this.backend, [["CircuitVerify", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'CircuitVerifyResponse') {
                throw new Error(`Expected variant name 'CircuitVerifyResponse' but got '${variantName}'`);
            }
            return toCircuitVerifyResponse(result);
        });
    }
    clientIvcComputeStandaloneVk(command) {
        const msgpackCommand = fromClientIvcComputeStandaloneVk(command);
        return msgpackCall(this.backend, [["ClientIvcComputeStandaloneVk", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'ClientIvcComputeStandaloneVkResponse') {
                throw new Error(`Expected variant name 'ClientIvcComputeStandaloneVkResponse' but got '${variantName}'`);
            }
            return toClientIvcComputeStandaloneVkResponse(result);
        });
    }
    clientIvcComputeIvcVk(command) {
        const msgpackCommand = fromClientIvcComputeIvcVk(command);
        return msgpackCall(this.backend, [["ClientIvcComputeIvcVk", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'ClientIvcComputeIvcVkResponse') {
                throw new Error(`Expected variant name 'ClientIvcComputeIvcVkResponse' but got '${variantName}'`);
            }
            return toClientIvcComputeIvcVkResponse(result);
        });
    }
    clientIvcStart(command) {
        const msgpackCommand = fromClientIvcStart(command);
        return msgpackCall(this.backend, [["ClientIvcStart", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'ClientIvcStartResponse') {
                throw new Error(`Expected variant name 'ClientIvcStartResponse' but got '${variantName}'`);
            }
            return toClientIvcStartResponse(result);
        });
    }
    clientIvcLoad(command) {
        const msgpackCommand = fromClientIvcLoad(command);
        return msgpackCall(this.backend, [["ClientIvcLoad", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'ClientIvcLoadResponse') {
                throw new Error(`Expected variant name 'ClientIvcLoadResponse' but got '${variantName}'`);
            }
            return toClientIvcLoadResponse(result);
        });
    }
    clientIvcAccumulate(command) {
        const msgpackCommand = fromClientIvcAccumulate(command);
        return msgpackCall(this.backend, [["ClientIvcAccumulate", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'ClientIvcAccumulateResponse') {
                throw new Error(`Expected variant name 'ClientIvcAccumulateResponse' but got '${variantName}'`);
            }
            return toClientIvcAccumulateResponse(result);
        });
    }
    clientIvcProve(command) {
        const msgpackCommand = fromClientIvcProve(command);
        return msgpackCall(this.backend, [["ClientIvcProve", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'ClientIvcProveResponse') {
                throw new Error(`Expected variant name 'ClientIvcProveResponse' but got '${variantName}'`);
            }
            return toClientIvcProveResponse(result);
        });
    }
    clientIvcVerify(command) {
        const msgpackCommand = fromClientIvcVerify(command);
        return msgpackCall(this.backend, [["ClientIvcVerify", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'ClientIvcVerifyResponse') {
                throw new Error(`Expected variant name 'ClientIvcVerifyResponse' but got '${variantName}'`);
            }
            return toClientIvcVerifyResponse(result);
        });
    }
    vkAsFields(command) {
        const msgpackCommand = fromVkAsFields(command);
        return msgpackCall(this.backend, [["VkAsFields", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'VkAsFieldsResponse') {
                throw new Error(`Expected variant name 'VkAsFieldsResponse' but got '${variantName}'`);
            }
            return toVkAsFieldsResponse(result);
        });
    }
    megaVkAsFields(command) {
        const msgpackCommand = fromMegaVkAsFields(command);
        return msgpackCall(this.backend, [["MegaVkAsFields", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'MegaVkAsFieldsResponse') {
                throw new Error(`Expected variant name 'MegaVkAsFieldsResponse' but got '${variantName}'`);
            }
            return toMegaVkAsFieldsResponse(result);
        });
    }
    circuitWriteSolidityVerifier(command) {
        const msgpackCommand = fromCircuitWriteSolidityVerifier(command);
        return msgpackCall(this.backend, [["CircuitWriteSolidityVerifier", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'CircuitWriteSolidityVerifierResponse') {
                throw new Error(`Expected variant name 'CircuitWriteSolidityVerifierResponse' but got '${variantName}'`);
            }
            return toCircuitWriteSolidityVerifierResponse(result);
        });
    }
    clientIvcCheckPrecomputedVk(command) {
        const msgpackCommand = fromClientIvcCheckPrecomputedVk(command);
        return msgpackCall(this.backend, [["ClientIvcCheckPrecomputedVk", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'ClientIvcCheckPrecomputedVkResponse') {
                throw new Error(`Expected variant name 'ClientIvcCheckPrecomputedVkResponse' but got '${variantName}'`);
            }
            return toClientIvcCheckPrecomputedVkResponse(result);
        });
    }
    clientIvcStats(command) {
        const msgpackCommand = fromClientIvcStats(command);
        return msgpackCall(this.backend, [["ClientIvcStats", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'ClientIvcStatsResponse') {
                throw new Error(`Expected variant name 'ClientIvcStatsResponse' but got '${variantName}'`);
            }
            return toClientIvcStatsResponse(result);
        });
    }
    poseidon2Hash(command) {
        const msgpackCommand = fromPoseidon2Hash(command);
        return msgpackCall(this.backend, [["Poseidon2Hash", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'Poseidon2HashResponse') {
                throw new Error(`Expected variant name 'Poseidon2HashResponse' but got '${variantName}'`);
            }
            return toPoseidon2HashResponse(result);
        });
    }
    poseidon2Permutation(command) {
        const msgpackCommand = fromPoseidon2Permutation(command);
        return msgpackCall(this.backend, [["Poseidon2Permutation", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'Poseidon2PermutationResponse') {
                throw new Error(`Expected variant name 'Poseidon2PermutationResponse' but got '${variantName}'`);
            }
            return toPoseidon2PermutationResponse(result);
        });
    }
    poseidon2HashAccumulate(command) {
        const msgpackCommand = fromPoseidon2HashAccumulate(command);
        return msgpackCall(this.backend, [["Poseidon2HashAccumulate", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'Poseidon2HashAccumulateResponse') {
                throw new Error(`Expected variant name 'Poseidon2HashAccumulateResponse' but got '${variantName}'`);
            }
            return toPoseidon2HashAccumulateResponse(result);
        });
    }
    pedersenCommit(command) {
        const msgpackCommand = fromPedersenCommit(command);
        return msgpackCall(this.backend, [["PedersenCommit", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'PedersenCommitResponse') {
                throw new Error(`Expected variant name 'PedersenCommitResponse' but got '${variantName}'`);
            }
            return toPedersenCommitResponse(result);
        });
    }
    pedersenHash(command) {
        const msgpackCommand = fromPedersenHash(command);
        return msgpackCall(this.backend, [["PedersenHash", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'PedersenHashResponse') {
                throw new Error(`Expected variant name 'PedersenHashResponse' but got '${variantName}'`);
            }
            return toPedersenHashResponse(result);
        });
    }
    pedersenHashBuffer(command) {
        const msgpackCommand = fromPedersenHashBuffer(command);
        return msgpackCall(this.backend, [["PedersenHashBuffer", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'PedersenHashBufferResponse') {
                throw new Error(`Expected variant name 'PedersenHashBufferResponse' but got '${variantName}'`);
            }
            return toPedersenHashBufferResponse(result);
        });
    }
    blake2s(command) {
        const msgpackCommand = fromBlake2s(command);
        return msgpackCall(this.backend, [["Blake2s", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'Blake2sResponse') {
                throw new Error(`Expected variant name 'Blake2sResponse' but got '${variantName}'`);
            }
            return toBlake2sResponse(result);
        });
    }
    blake2sToField(command) {
        const msgpackCommand = fromBlake2sToField(command);
        return msgpackCall(this.backend, [["Blake2sToField", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'Blake2sToFieldResponse') {
                throw new Error(`Expected variant name 'Blake2sToFieldResponse' but got '${variantName}'`);
            }
            return toBlake2sToFieldResponse(result);
        });
    }
    aesEncrypt(command) {
        const msgpackCommand = fromAesEncrypt(command);
        return msgpackCall(this.backend, [["AesEncrypt", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'AesEncryptResponse') {
                throw new Error(`Expected variant name 'AesEncryptResponse' but got '${variantName}'`);
            }
            return toAesEncryptResponse(result);
        });
    }
    aesDecrypt(command) {
        const msgpackCommand = fromAesDecrypt(command);
        return msgpackCall(this.backend, [["AesDecrypt", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'AesDecryptResponse') {
                throw new Error(`Expected variant name 'AesDecryptResponse' but got '${variantName}'`);
            }
            return toAesDecryptResponse(result);
        });
    }
    grumpkinMul(command) {
        const msgpackCommand = fromGrumpkinMul(command);
        return msgpackCall(this.backend, [["GrumpkinMul", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'GrumpkinMulResponse') {
                throw new Error(`Expected variant name 'GrumpkinMulResponse' but got '${variantName}'`);
            }
            return toGrumpkinMulResponse(result);
        });
    }
    grumpkinAdd(command) {
        const msgpackCommand = fromGrumpkinAdd(command);
        return msgpackCall(this.backend, [["GrumpkinAdd", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'GrumpkinAddResponse') {
                throw new Error(`Expected variant name 'GrumpkinAddResponse' but got '${variantName}'`);
            }
            return toGrumpkinAddResponse(result);
        });
    }
    grumpkinBatchMul(command) {
        const msgpackCommand = fromGrumpkinBatchMul(command);
        return msgpackCall(this.backend, [["GrumpkinBatchMul", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'GrumpkinBatchMulResponse') {
                throw new Error(`Expected variant name 'GrumpkinBatchMulResponse' but got '${variantName}'`);
            }
            return toGrumpkinBatchMulResponse(result);
        });
    }
    grumpkinGetRandomFr(command) {
        const msgpackCommand = fromGrumpkinGetRandomFr(command);
        return msgpackCall(this.backend, [["GrumpkinGetRandomFr", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'GrumpkinGetRandomFrResponse') {
                throw new Error(`Expected variant name 'GrumpkinGetRandomFrResponse' but got '${variantName}'`);
            }
            return toGrumpkinGetRandomFrResponse(result);
        });
    }
    grumpkinReduce512(command) {
        const msgpackCommand = fromGrumpkinReduce512(command);
        return msgpackCall(this.backend, [["GrumpkinReduce512", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'GrumpkinReduce512Response') {
                throw new Error(`Expected variant name 'GrumpkinReduce512Response' but got '${variantName}'`);
            }
            return toGrumpkinReduce512Response(result);
        });
    }
    secp256k1Mul(command) {
        const msgpackCommand = fromSecp256k1Mul(command);
        return msgpackCall(this.backend, [["Secp256k1Mul", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'Secp256k1MulResponse') {
                throw new Error(`Expected variant name 'Secp256k1MulResponse' but got '${variantName}'`);
            }
            return toSecp256k1MulResponse(result);
        });
    }
    secp256k1GetRandomFr(command) {
        const msgpackCommand = fromSecp256k1GetRandomFr(command);
        return msgpackCall(this.backend, [["Secp256k1GetRandomFr", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'Secp256k1GetRandomFrResponse') {
                throw new Error(`Expected variant name 'Secp256k1GetRandomFrResponse' but got '${variantName}'`);
            }
            return toSecp256k1GetRandomFrResponse(result);
        });
    }
    secp256k1Reduce512(command) {
        const msgpackCommand = fromSecp256k1Reduce512(command);
        return msgpackCall(this.backend, [["Secp256k1Reduce512", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'Secp256k1Reduce512Response') {
                throw new Error(`Expected variant name 'Secp256k1Reduce512Response' but got '${variantName}'`);
            }
            return toSecp256k1Reduce512Response(result);
        });
    }
    bn254FrSqrt(command) {
        const msgpackCommand = fromBn254FrSqrt(command);
        return msgpackCall(this.backend, [["Bn254FrSqrt", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'Bn254FrSqrtResponse') {
                throw new Error(`Expected variant name 'Bn254FrSqrtResponse' but got '${variantName}'`);
            }
            return toBn254FrSqrtResponse(result);
        });
    }
    schnorrComputePublicKey(command) {
        const msgpackCommand = fromSchnorrComputePublicKey(command);
        return msgpackCall(this.backend, [["SchnorrComputePublicKey", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'SchnorrComputePublicKeyResponse') {
                throw new Error(`Expected variant name 'SchnorrComputePublicKeyResponse' but got '${variantName}'`);
            }
            return toSchnorrComputePublicKeyResponse(result);
        });
    }
    schnorrConstructSignature(command) {
        const msgpackCommand = fromSchnorrConstructSignature(command);
        return msgpackCall(this.backend, [["SchnorrConstructSignature", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'SchnorrConstructSignatureResponse') {
                throw new Error(`Expected variant name 'SchnorrConstructSignatureResponse' but got '${variantName}'`);
            }
            return toSchnorrConstructSignatureResponse(result);
        });
    }
    schnorrVerifySignature(command) {
        const msgpackCommand = fromSchnorrVerifySignature(command);
        return msgpackCall(this.backend, [["SchnorrVerifySignature", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'SchnorrVerifySignatureResponse') {
                throw new Error(`Expected variant name 'SchnorrVerifySignatureResponse' but got '${variantName}'`);
            }
            return toSchnorrVerifySignatureResponse(result);
        });
    }
    ecdsaSecp256k1ComputePublicKey(command) {
        const msgpackCommand = fromEcdsaSecp256k1ComputePublicKey(command);
        return msgpackCall(this.backend, [["EcdsaSecp256k1ComputePublicKey", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'EcdsaSecp256k1ComputePublicKeyResponse') {
                throw new Error(`Expected variant name 'EcdsaSecp256k1ComputePublicKeyResponse' but got '${variantName}'`);
            }
            return toEcdsaSecp256k1ComputePublicKeyResponse(result);
        });
    }
    ecdsaSecp256r1ComputePublicKey(command) {
        const msgpackCommand = fromEcdsaSecp256r1ComputePublicKey(command);
        return msgpackCall(this.backend, [["EcdsaSecp256r1ComputePublicKey", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'EcdsaSecp256r1ComputePublicKeyResponse') {
                throw new Error(`Expected variant name 'EcdsaSecp256r1ComputePublicKeyResponse' but got '${variantName}'`);
            }
            return toEcdsaSecp256r1ComputePublicKeyResponse(result);
        });
    }
    ecdsaSecp256k1ConstructSignature(command) {
        const msgpackCommand = fromEcdsaSecp256k1ConstructSignature(command);
        return msgpackCall(this.backend, [["EcdsaSecp256k1ConstructSignature", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'EcdsaSecp256k1ConstructSignatureResponse') {
                throw new Error(`Expected variant name 'EcdsaSecp256k1ConstructSignatureResponse' but got '${variantName}'`);
            }
            return toEcdsaSecp256k1ConstructSignatureResponse(result);
        });
    }
    ecdsaSecp256r1ConstructSignature(command) {
        const msgpackCommand = fromEcdsaSecp256r1ConstructSignature(command);
        return msgpackCall(this.backend, [["EcdsaSecp256r1ConstructSignature", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'EcdsaSecp256r1ConstructSignatureResponse') {
                throw new Error(`Expected variant name 'EcdsaSecp256r1ConstructSignatureResponse' but got '${variantName}'`);
            }
            return toEcdsaSecp256r1ConstructSignatureResponse(result);
        });
    }
    ecdsaSecp256k1RecoverPublicKey(command) {
        const msgpackCommand = fromEcdsaSecp256k1RecoverPublicKey(command);
        return msgpackCall(this.backend, [["EcdsaSecp256k1RecoverPublicKey", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'EcdsaSecp256k1RecoverPublicKeyResponse') {
                throw new Error(`Expected variant name 'EcdsaSecp256k1RecoverPublicKeyResponse' but got '${variantName}'`);
            }
            return toEcdsaSecp256k1RecoverPublicKeyResponse(result);
        });
    }
    ecdsaSecp256r1RecoverPublicKey(command) {
        const msgpackCommand = fromEcdsaSecp256r1RecoverPublicKey(command);
        return msgpackCall(this.backend, [["EcdsaSecp256r1RecoverPublicKey", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'EcdsaSecp256r1RecoverPublicKeyResponse') {
                throw new Error(`Expected variant name 'EcdsaSecp256r1RecoverPublicKeyResponse' but got '${variantName}'`);
            }
            return toEcdsaSecp256r1RecoverPublicKeyResponse(result);
        });
    }
    ecdsaSecp256k1VerifySignature(command) {
        const msgpackCommand = fromEcdsaSecp256k1VerifySignature(command);
        return msgpackCall(this.backend, [["EcdsaSecp256k1VerifySignature", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'EcdsaSecp256k1VerifySignatureResponse') {
                throw new Error(`Expected variant name 'EcdsaSecp256k1VerifySignatureResponse' but got '${variantName}'`);
            }
            return toEcdsaSecp256k1VerifySignatureResponse(result);
        });
    }
    ecdsaSecp256r1VerifySignature(command) {
        const msgpackCommand = fromEcdsaSecp256r1VerifySignature(command);
        return msgpackCall(this.backend, [["EcdsaSecp256r1VerifySignature", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'EcdsaSecp256r1VerifySignatureResponse') {
                throw new Error(`Expected variant name 'EcdsaSecp256r1VerifySignatureResponse' but got '${variantName}'`);
            }
            return toEcdsaSecp256r1VerifySignatureResponse(result);
        });
    }
    srsInitSrs(command) {
        const msgpackCommand = fromSrsInitSrs(command);
        return msgpackCall(this.backend, [["SrsInitSrs", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'SrsInitSrsResponse') {
                throw new Error(`Expected variant name 'SrsInitSrsResponse' but got '${variantName}'`);
            }
            return toSrsInitSrsResponse(result);
        });
    }
    srsInitGrumpkinSrs(command) {
        const msgpackCommand = fromSrsInitGrumpkinSrs(command);
        return msgpackCall(this.backend, [["SrsInitGrumpkinSrs", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'SrsInitGrumpkinSrsResponse') {
                throw new Error(`Expected variant name 'SrsInitGrumpkinSrsResponse' but got '${variantName}'`);
            }
            return toSrsInitGrumpkinSrsResponse(result);
        });
    }
    shutdown(command) {
        const msgpackCommand = fromShutdown(command);
        return msgpackCall(this.backend, [["Shutdown", msgpackCommand]]).then(([variantName, result]) => {
            if (variantName !== 'ShutdownResponse') {
                throw new Error(`Expected variant name 'ShutdownResponse' but got '${variantName}'`);
            }
            return toShutdownResponse(result);
        });
    }
    destroy() {
        return this.backend.destroy ? this.backend.destroy() : Promise.resolve();
    }
}
