// @aztec/bb.js/dest/node-cjs/barretenberg/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BarretenbergSync = exports.Barretenberg = exports.AztecClientBackend = exports.UltraHonkVerifierBackend = exports.UltraHonkBackend = void 0;
const tslib_1 = require("tslib");
const index_js_1 = require("../crs/index.js");
const index_js_2 = require("../log/index.js");
const async_js_1 = require("../cbind/generated/async.js");
const sync_js_1 = require("../cbind/generated/sync.js");
const index_js_3 = require("../bb_backends/index.js");
const index_js_4 = require("../bb_backends/node/index.js");
var backend_js_1 = require("./backend.js");
Object.defineProperty(exports, "UltraHonkBackend", { enumerable: true, get: function () { return backend_js_1.UltraHonkBackend; } });
Object.defineProperty(exports, "UltraHonkVerifierBackend", { enumerable: true, get: function () { return backend_js_1.UltraHonkVerifierBackend; } });
Object.defineProperty(exports, "AztecClientBackend", { enumerable: true, get: function () { return backend_js_1.AztecClientBackend; } });
tslib_1.__exportStar(require("../bb_backends/index.js"), exports);
/**
 * The main class library consumers interact with.
 * It extends the generated api, and provides a static constructor "new" to compose components.
 */
class Barretenberg extends async_js_1.AsyncApi {
    constructor(backend, options) {
        super(backend);
        this.options = options;
    }
    /**
     * Constructs an instance of Barretenberg.
     *
     * If options.backend is set: uses that specific backend (throws if unavailable)
     * If options.backend is unset: tries backends in order with fallback:
     *   1. NativeSharedMemory (if bb binary available)
     *   2. WasmWorker (in browser) or Wasm (in Node.js)
     */
    static async new(options = {}) {
        const logger = options.logger ?? (0, index_js_2.createDebugLogger)('bb_async');
        if (options.backend) {
            // Explicit backend required - no fallback
            return await (0, index_js_4.createAsyncBackend)(options.backend, options, logger);
        }
        if (typeof window === 'undefined') {
            try {
                return await (0, index_js_4.createAsyncBackend)(index_js_3.BackendType.NativeSharedMemory, options, logger);
            }
            catch (err) {
                logger(`Shared memory unavailable (${err.message}), falling back to other backends`);
                try {
                    return await (0, index_js_4.createAsyncBackend)(index_js_3.BackendType.NativeUnixSocket, options, logger);
                }
                catch (err) {
                    logger(`Unix socket unavailable (${err.message}), falling back to WASM`);
                    return await (0, index_js_4.createAsyncBackend)(index_js_3.BackendType.Wasm, options, logger);
                }
            }
        }
        else {
            logger(`In browser, using WASM over worker backend.`);
            return await (0, index_js_4.createAsyncBackend)(index_js_3.BackendType.WasmWorker, options, logger);
        }
    }
    async initSRSForCircuitSize(circuitSize) {
        const minSRSSize = 2 ** 9; // 2**9 is the dyadic size for the SmallSubgroupIPA MSM.
        const crs = await index_js_1.Crs.new(Math.max(circuitSize, minSRSSize) + 1, this.options.crsPath, this.options.logger);
        // TODO(https://github.com/AztecProtocol/barretenberg/issues/1129): Do slab allocator initialization?
        // await this.commonInitSlabAllocator(circuitSize);
        await this.srsInitSrs({ pointsBuf: crs.getG1Data(), numPoints: crs.numPoints, g2Point: crs.getG2Data() });
    }
    async initSRSClientIVC(srsSize = this.getDefaultSrsSize()) {
        // crsPath can be undefined
        const crs = await index_js_1.Crs.new(srsSize + 1, this.options.crsPath, this.options.logger);
        const grumpkinCrs = await index_js_1.GrumpkinCrs.new(2 ** 16 + 1, this.options.crsPath, this.options.logger);
        // Load CRS into wasm global CRS state.
        // TODO: Make RawBuffer be default behavior, and have a specific Vector type for when wanting length prefixed.
        await this.srsInitSrs({ pointsBuf: crs.getG1Data(), numPoints: crs.numPoints, g2Point: crs.getG2Data() });
        await this.srsInitGrumpkinSrs({ pointsBuf: grumpkinCrs.getG1Data(), numPoints: grumpkinCrs.numPoints });
    }
    getDefaultSrsSize() {
        // iOS browser is very aggressive with memory. Check if running in browser and on iOS
        // We expect the mobile iOS browser to kill us >=1GB, so no real use in using a larger SRS.
        if (typeof window !== 'undefined' && /iPad|iPhone/.test(navigator.userAgent)) {
            return 2 ** 18;
        }
        return 2 ** 20;
    }
    async acirGetCircuitSizes(bytecode, recursive, honkRecursion) {
        const response = await this.circuitStats({
            circuit: { name: '', bytecode, verificationKey: new Uint8Array() },
            includeGatesPerOpcode: false,
            settings: {
                ipaAccumulation: false,
                oracleHashType: honkRecursion ? 'poseidon2' : 'keccak',
                disableZk: !recursive,
                optimizedSolidityVerifier: false,
            },
        });
        return [response.numGates, response.numGatesDyadic];
    }
    async acirInitSRS(bytecode, recursive, honkRecursion) {
        const [_, subgroupSize] = await this.acirGetCircuitSizes(bytecode, recursive, honkRecursion);
        return this.initSRSForCircuitSize(subgroupSize);
    }
    async destroy() {
        return super.destroy();
    }
    /**
     * Initialize the singleton instance of Barretenberg.
     * @param options Backend configuration options
     */
    static async initSingleton(options = {}) {
        if (!barretenbergSingletonPromise) {
            barretenbergSingletonPromise = Barretenberg.new(options);
        }
        try {
            barretenbergSingleton = await barretenbergSingletonPromise;
            return barretenbergSingleton;
        }
        catch (error) {
            // If initialization fails, clear the singleton so next call can retry
            barretenbergSingleton = undefined;
            barretenbergSingletonPromise = undefined;
            throw error;
        }
    }
    static async destroySingleton() {
        if (barretenbergSingleton) {
            await barretenbergSingleton.destroy();
            barretenbergSingleton = undefined;
            barretenbergSingletonPromise = undefined;
        }
    }
    /**
     * Get the singleton instance of Barretenberg.
     * Must call initSingleton() first.
     */
    static getSingleton() {
        if (!barretenbergSingleton) {
            throw new Error('First call Barretenberg.initSingleton() on @aztec/bb.js module.');
        }
        return barretenbergSingleton;
    }
}
exports.Barretenberg = Barretenberg;
let barretenbergSingletonPromise;
let barretenbergSingleton;
let barretenbergSyncSingletonPromise;
let barretenbergSyncSingleton;
class BarretenbergSync extends sync_js_1.SyncApi {
    constructor(backend) {
        super(backend);
    }
    /**
     * Create a new BarretenbergSync instance.
     *
     * If options.backend is set: uses that specific backend (throws if unavailable)
     * If options.backend is unset: tries backends in order with fallback:
     *   1. NativeSharedMem (if bb binary + NAPI module available)
     *   2. Wasm
     *
     * Supported backends: Wasm, NativeSharedMem
     * Not supported: WasmWorker (no workers in sync), NativeUnixSocket (async only)
     */
    static async new(options = {}) {
        const logger = options.logger ?? (0, index_js_2.createDebugLogger)('bb_sync');
        if (options.backend) {
            return await (0, index_js_4.createSyncBackend)(options.backend, options, logger);
        }
        // Try native, fallback to WASM.
        try {
            return await (0, index_js_4.createSyncBackend)(index_js_3.BackendType.NativeSharedMemory, options, logger);
        }
        catch (err) {
            logger(`Shared memory unavailable (${err.message}), falling back to WASM`);
        }
        return await (0, index_js_4.createSyncBackend)(index_js_3.BackendType.Wasm, options, logger);
    }
    /**
     * Initialize the singleton instance.
     * @param options Backend configuration options
     */
    static async initSingleton(options = {}) {
        if (!barretenbergSyncSingletonPromise) {
            barretenbergSyncSingletonPromise = BarretenbergSync.new(options);
        }
        barretenbergSyncSingleton = await barretenbergSyncSingletonPromise;
        return barretenbergSyncSingleton;
    }
    static destroySingleton() {
        if (barretenbergSyncSingleton) {
            barretenbergSyncSingleton.destroy();
            barretenbergSyncSingleton = undefined;
            barretenbergSyncSingletonPromise = undefined;
        }
    }
    static getSingleton() {
        if (!barretenbergSyncSingleton) {
            throw new Error('First call BarretenbergSync.initSingleton() on @aztec/bb.js module.');
        }
        return barretenbergSyncSingleton;
    }
}
exports.BarretenbergSync = BarretenbergSync;
