// @aztec/aztec.js/dest/contract/base_contract_interaction.js
import { createLogger } from '@aztec/foundation/log';
import { toSendOptions } from './interaction_options.js';
import { SentTx } from './sent_tx.js';
/**
 * Base class for an interaction with a contract, be it a deployment, a function call, or a batch.
 * Implements the sequence create/simulate/send.
 */ export class BaseContractInteraction {
    wallet;
    authWitnesses;
    capsules;
    log;
    constructor(wallet, authWitnesses = [], capsules = []){
        this.wallet = wallet;
        this.authWitnesses = authWitnesses;
        this.capsules = capsules;
        this.log = createLogger('aztecjs:contract_interaction');
    }
    // docs:start:send
    /**
   * Sends a transaction to the contract function with the specified options.
   * This function throws an error if called on a utility function.
   * It creates and signs the transaction if necessary, and returns a SentTx instance,
   * which can be used to track the transaction status, receipt, and events.
   * @param options - An object containing 'from' property representing
   * the AztecAddress of the sender and optional fee configuration
   * @returns A SentTx instance for tracking the transaction status and information.
   */ send(options) {
        // docs:end:send
        const sendTx = async ()=>{
            const executionPayload = await this.request(options);
            const sendOptions = await toSendOptions(options);
            return this.wallet.sendTx(executionPayload, sendOptions);
        };
        return new SentTx(this.wallet, sendTx);
    }
}
