// @aztec/aztec.js/dest/contract/interaction_options.js
/**
 * Transforms and cleans up the higher level SendInteractionOptions defined by the interaction into
 * SendOptions, which are the ones that can be serialized and forwarded to the wallet
 */ export async function toSendOptions(options) {
    return {
        ...options,
        fee: {
            // If this interaction includes a fee payment method, pass the fee payer
            // as a hint to the wallet
            embeddedPaymentMethodFeePayer: await options.fee?.paymentMethod?.getFeePayer(),
            // If a payment method that includes gas settings was used,
            // try to reuse as much as possible while still allowing
            // manual override. CAREFUL: this can cause mismatches during proving
            gasSettings: {
                ...options.fee?.paymentMethod?.getGasSettings(),
                ...options.fee?.gasSettings
            }
        }
    };
}
/**
 * Transforms and cleans up the higher level SimulateInteractionOptions defined by the interaction into
 * SimulateOptions, which are the ones that can be serialized and forwarded to the wallet
 */ export async function toSimulateOptions(options) {
    return {
        ...options,
        fee: {
            // If this interaction includes a fee payment method, pass the fee payer
            // as a hint to the wallet
            embeddedPaymentMethodFeePayer: await options.fee?.paymentMethod?.getFeePayer(),
            // If a payment method that includes gas settings was used,
            // try to reuse as much as possible while still allowing
            // manual override. CAREFUL: this can cause mismatches during proving
            gasSettings: {
                ...options.fee?.paymentMethod?.getGasSettings(),
                ...options.fee?.gasSettings
            },
            estimateGas: options.fee?.estimateGas,
            estimatedGasPadding: options.fee?.estimatedGasPadding
        }
    };
}
/**
 * Transforms and cleans up the higher level ProfileInteractionOptions defined by the interaction into
 * ProfileOptions, which are the ones that can be serialized and forwarded to the wallet
 */ export async function toProfileOptions(options) {
    return {
        ...await toSimulateOptions(options),
        profileMode: options.profileMode,
        skipProofGeneration: options.skipProofGeneration
    };
}
