// @aztec/protocol-contracts/src/protocol_contract_data.ts

    // GENERATED FILE - DO NOT EDIT. RUN `yarn generate` or `yarn generate:data`
    import { Fr } from '@aztec/foundation/fields';
    import { AztecAddress } from '@aztec/stdlib/aztec-address';
    import { ProtocolContracts } from '@aztec/stdlib/tx';

    
    export const protocolContractNames = [
      'AuthRegistry',
'ContractInstanceRegistry',
'ContractClassRegistry',
'MultiCallEntrypoint',
'FeeJuice',
'Router'
    ] as const;

    export type ProtocolContractName = typeof protocolContractNames[number];
  

    
    export const ProtocolContractSalt: Record<ProtocolContractName, Fr> = {
      AuthRegistry: new Fr(1),
ContractInstanceRegistry: new Fr(1),
ContractClassRegistry: new Fr(1),
MultiCallEntrypoint: new Fr(1),
FeeJuice: new Fr(1),
Router: new Fr(1)
    };
  

    
    export const ProtocolContractAddress: Record<ProtocolContractName, AztecAddress> = {
      AuthRegistry: AztecAddress.fromBigInt(1n),
ContractInstanceRegistry: AztecAddress.fromBigInt(2n),
ContractClassRegistry: AztecAddress.fromBigInt(3n),
MultiCallEntrypoint: AztecAddress.fromBigInt(4n),
FeeJuice: AztecAddress.fromBigInt(5n),
Router: AztecAddress.fromBigInt(6n)
    };
  

    
    export const ProtocolContractDerivedAddress = {
      AuthRegistry: AztecAddress.fromString('0x211f1e41d3063b70fbd60030ffcee7020a405d4777af3467f20b0d7fe5637a61'),
ContractInstanceRegistry: AztecAddress.fromString('0x1f33581fd9e68afaa9afb20370a2059911fa816b5d7a348903555c06b31d79a5'),
ContractClassRegistry: AztecAddress.fromString('0x25b3403229608574dd1397592093c27e7f1b37018c199e4c00608010c404c490'),
MultiCallEntrypoint: AztecAddress.fromString('0x016c38dc8b63d9589f872fa9ed121c0700e1cc3c695e540349e8e9b727c1233a'),
FeeJuice: AztecAddress.fromString('0x1363751f5f38916a38a88d8b536c53eaf07058e077b22bef4288b0f9178021dc'),
Router: AztecAddress.fromString('0x17a6c8d0be66035edb32677e3f121669926df68bef5b009024216b3c22d60183')
    };
  

    
    export const ProtocolContractsList = new ProtocolContracts([
      AztecAddress.fromString('0x211f1e41d3063b70fbd60030ffcee7020a405d4777af3467f20b0d7fe5637a61'),
AztecAddress.fromString('0x1f33581fd9e68afaa9afb20370a2059911fa816b5d7a348903555c06b31d79a5'),
AztecAddress.fromString('0x25b3403229608574dd1397592093c27e7f1b37018c199e4c00608010c404c490'),
AztecAddress.fromString('0x016c38dc8b63d9589f872fa9ed121c0700e1cc3c695e540349e8e9b727c1233a'),
AztecAddress.fromString('0x1363751f5f38916a38a88d8b536c53eaf07058e077b22bef4288b0f9178021dc'),
AztecAddress.fromString('0x17a6c8d0be66035edb32677e3f121669926df68bef5b009024216b3c22d60183'),
AztecAddress.fromString('0x0000000000000000000000000000000000000000000000000000000000000000'),
AztecAddress.fromString('0x0000000000000000000000000000000000000000000000000000000000000000'),
AztecAddress.fromString('0x0000000000000000000000000000000000000000000000000000000000000000'),
AztecAddress.fromString('0x0000000000000000000000000000000000000000000000000000000000000000'),
AztecAddress.fromString('0x0000000000000000000000000000000000000000000000000000000000000000')
    ]);

    export const protocolContractsHash = Fr.fromString('0x27638bf36029626c609c64210f0e027e95b2975edc8359afa6f7a77974acf52a');
  

    
  export const CONTRACT_INSTANCE_PUBLISHED_EVENT_TAG = Fr.fromHexString('0x28da3fc471387d4e20b646b77abc8cdfc02a339a7507285a73d84f9b97b13486');
