// @aztec/accounts/dest/defaults/account_interface.js
import { DefaultAccountEntrypoint } from '@aztec/entrypoints/account';
/**
 * Default implementation for an account interface. Requires that the account uses the default
 * entrypoint signature, which accept an AppPayload and a FeePayload as defined in noir-libs/aztec-noir/src/entrypoint module
 */ export class DefaultAccountInterface {
    authWitnessProvider;
    address;
    entrypoint;
    chainId;
    version;
    constructor(authWitnessProvider, address, chainInfo){
        this.authWitnessProvider = authWitnessProvider;
        this.address = address;
        this.entrypoint = new DefaultAccountEntrypoint(address.address, authWitnessProvider, chainInfo.chainId.toNumber(), chainInfo.version.toNumber());
        this.chainId = chainInfo.chainId;
        this.version = chainInfo.version;
    }
    createTxExecutionRequest(exec, gasSettings, options) {
        return this.entrypoint.createTxExecutionRequest(exec, gasSettings, options);
    }
    createAuthWit(messageHash) {
        return this.authWitnessProvider.createAuthWit(messageHash);
    }
    getCompleteAddress() {
        return this.address;
    }
    getAddress() {
        return this.address.address;
    }
    getChainId() {
        return this.chainId;
    }
    getVersion() {
        return this.version;
    }
}
