// @aztec/aztec.js/dest/account/signerless_account.js
import { DefaultMultiCallEntrypoint } from '@aztec/entrypoints/multicall';
/**
 * Account implementation which creates a transaction using the multicall protocol contract as entrypoint.
 */ export class SignerlessAccount {
    entrypoint;
    constructor(chainInfo){
        this.entrypoint = new DefaultMultiCallEntrypoint(chainInfo.chainId.toNumber(), chainInfo.version.toNumber());
    }
    createTxExecutionRequest(exec, gasSettings) {
        return this.entrypoint.createTxExecutionRequest(exec, gasSettings);
    }
    getChainId() {
        throw new Error('SignerlessAccount: Method getChainId not implemented.');
    }
    getVersion() {
        throw new Error('SignerlessAccount: Method getVersion not implemented.');
    }
    getCompleteAddress() {
        throw new Error('SignerlessAccount: Method getCompleteAddress not implemented.');
    }
    getAddress() {
        throw new Error('SignerlessAccount: Method getAddress not implemented.');
    }
    createAuthWit(_intent) {
        throw new Error('SignerlessAccount: Method createAuthWit not implemented.');
    }
}
