// @aztec/bb.js/dest/node-cjs/cbind/generated/api_types.js
"use strict";
// AUTOGENERATED FILE - DO NOT EDIT
Object.defineProperty(exports, "__esModule", { value: true });
exports.toAesDecrypt = toAesDecrypt;
exports.fromAesDecrypt = fromAesDecrypt;
exports.toAesDecryptResponse = toAesDecryptResponse;
exports.fromAesDecryptResponse = fromAesDecryptResponse;
exports.toAesEncrypt = toAesEncrypt;
exports.fromAesEncrypt = fromAesEncrypt;
exports.toAesEncryptResponse = toAesEncryptResponse;
exports.fromAesEncryptResponse = fromAesEncryptResponse;
exports.toAffineElement = toAffineElement;
exports.fromAffineElement = fromAffineElement;
exports.toBlake2s = toBlake2s;
exports.fromBlake2s = fromBlake2s;
exports.toBlake2sResponse = toBlake2sResponse;
exports.fromBlake2sResponse = fromBlake2sResponse;
exports.toBlake2sToField = toBlake2sToField;
exports.fromBlake2sToField = fromBlake2sToField;
exports.toBlake2sToFieldResponse = toBlake2sToFieldResponse;
exports.fromBlake2sToFieldResponse = fromBlake2sToFieldResponse;
exports.toBn254FrSqrt = toBn254FrSqrt;
exports.fromBn254FrSqrt = fromBn254FrSqrt;
exports.toBn254FrSqrtResponse = toBn254FrSqrtResponse;
exports.fromBn254FrSqrtResponse = fromBn254FrSqrtResponse;
exports.toCircuitComputeVk = toCircuitComputeVk;
exports.fromCircuitComputeVk = fromCircuitComputeVk;
exports.toCircuitComputeVkResponse = toCircuitComputeVkResponse;
exports.fromCircuitComputeVkResponse = fromCircuitComputeVkResponse;
exports.toCircuitInfoResponse = toCircuitInfoResponse;
exports.fromCircuitInfoResponse = fromCircuitInfoResponse;
exports.toCircuitInput = toCircuitInput;
exports.fromCircuitInput = fromCircuitInput;
exports.toCircuitInputNoVK = toCircuitInputNoVK;
exports.fromCircuitInputNoVK = fromCircuitInputNoVK;
exports.toCircuitProve = toCircuitProve;
exports.fromCircuitProve = fromCircuitProve;
exports.toCircuitProveResponse = toCircuitProveResponse;
exports.fromCircuitProveResponse = fromCircuitProveResponse;
exports.toCircuitStats = toCircuitStats;
exports.fromCircuitStats = fromCircuitStats;
exports.toCircuitVerify = toCircuitVerify;
exports.fromCircuitVerify = fromCircuitVerify;
exports.toCircuitVerifyResponse = toCircuitVerifyResponse;
exports.fromCircuitVerifyResponse = fromCircuitVerifyResponse;
exports.toCircuitWriteSolidityVerifier = toCircuitWriteSolidityVerifier;
exports.fromCircuitWriteSolidityVerifier = fromCircuitWriteSolidityVerifier;
exports.toCircuitWriteSolidityVerifierResponse = toCircuitWriteSolidityVerifierResponse;
exports.fromCircuitWriteSolidityVerifierResponse = fromCircuitWriteSolidityVerifierResponse;
exports.toClientIvcAccumulate = toClientIvcAccumulate;
exports.fromClientIvcAccumulate = fromClientIvcAccumulate;
exports.toClientIvcAccumulateResponse = toClientIvcAccumulateResponse;
exports.fromClientIvcAccumulateResponse = fromClientIvcAccumulateResponse;
exports.toClientIvcCheckPrecomputedVk = toClientIvcCheckPrecomputedVk;
exports.fromClientIvcCheckPrecomputedVk = fromClientIvcCheckPrecomputedVk;
exports.toClientIvcCheckPrecomputedVkResponse = toClientIvcCheckPrecomputedVkResponse;
exports.fromClientIvcCheckPrecomputedVkResponse = fromClientIvcCheckPrecomputedVkResponse;
exports.toClientIvcComputeIvcVk = toClientIvcComputeIvcVk;
exports.fromClientIvcComputeIvcVk = fromClientIvcComputeIvcVk;
exports.toClientIvcComputeIvcVkResponse = toClientIvcComputeIvcVkResponse;
exports.fromClientIvcComputeIvcVkResponse = fromClientIvcComputeIvcVkResponse;
exports.toClientIvcComputeStandaloneVk = toClientIvcComputeStandaloneVk;
exports.fromClientIvcComputeStandaloneVk = fromClientIvcComputeStandaloneVk;
exports.toClientIvcComputeStandaloneVkResponse = toClientIvcComputeStandaloneVkResponse;
exports.fromClientIvcComputeStandaloneVkResponse = fromClientIvcComputeStandaloneVkResponse;
exports.toClientIvcLoad = toClientIvcLoad;
exports.fromClientIvcLoad = fromClientIvcLoad;
exports.toClientIvcLoadResponse = toClientIvcLoadResponse;
exports.fromClientIvcLoadResponse = fromClientIvcLoadResponse;
exports.toClientIVCProof = toClientIVCProof;
exports.fromClientIVCProof = fromClientIVCProof;
exports.toClientIvcProve = toClientIvcProve;
exports.fromClientIvcProve = fromClientIvcProve;
exports.toClientIvcProveResponse = toClientIvcProveResponse;
exports.fromClientIvcProveResponse = fromClientIvcProveResponse;
exports.toClientIvcStart = toClientIvcStart;
exports.fromClientIvcStart = fromClientIvcStart;
exports.toClientIvcStartResponse = toClientIvcStartResponse;
exports.fromClientIvcStartResponse = fromClientIvcStartResponse;
exports.toClientIvcStats = toClientIvcStats;
exports.fromClientIvcStats = fromClientIvcStats;
exports.toClientIvcStatsResponse = toClientIvcStatsResponse;
exports.fromClientIvcStatsResponse = fromClientIvcStatsResponse;
exports.toClientIvcVerify = toClientIvcVerify;
exports.fromClientIvcVerify = fromClientIvcVerify;
exports.toClientIvcVerifyResponse = toClientIvcVerifyResponse;
exports.fromClientIvcVerifyResponse = fromClientIvcVerifyResponse;
exports.toECCVMProof = toECCVMProof;
exports.fromECCVMProof = fromECCVMProof;
exports.toEcdsaSecp256k1ComputePublicKey = toEcdsaSecp256k1ComputePublicKey;
exports.fromEcdsaSecp256k1ComputePublicKey = fromEcdsaSecp256k1ComputePublicKey;
exports.toEcdsaSecp256k1ComputePublicKeyResponse = toEcdsaSecp256k1ComputePublicKeyResponse;
exports.fromEcdsaSecp256k1ComputePublicKeyResponse = fromEcdsaSecp256k1ComputePublicKeyResponse;
exports.toEcdsaSecp256k1ConstructSignature = toEcdsaSecp256k1ConstructSignature;
exports.fromEcdsaSecp256k1ConstructSignature = fromEcdsaSecp256k1ConstructSignature;
exports.toEcdsaSecp256k1ConstructSignatureResponse = toEcdsaSecp256k1ConstructSignatureResponse;
exports.fromEcdsaSecp256k1ConstructSignatureResponse = fromEcdsaSecp256k1ConstructSignatureResponse;
exports.toEcdsaSecp256k1RecoverPublicKey = toEcdsaSecp256k1RecoverPublicKey;
exports.fromEcdsaSecp256k1RecoverPublicKey = fromEcdsaSecp256k1RecoverPublicKey;
exports.toEcdsaSecp256k1RecoverPublicKeyResponse = toEcdsaSecp256k1RecoverPublicKeyResponse;
exports.fromEcdsaSecp256k1RecoverPublicKeyResponse = fromEcdsaSecp256k1RecoverPublicKeyResponse;
exports.toEcdsaSecp256k1VerifySignature = toEcdsaSecp256k1VerifySignature;
exports.fromEcdsaSecp256k1VerifySignature = fromEcdsaSecp256k1VerifySignature;
exports.toEcdsaSecp256k1VerifySignatureResponse = toEcdsaSecp256k1VerifySignatureResponse;
exports.fromEcdsaSecp256k1VerifySignatureResponse = fromEcdsaSecp256k1VerifySignatureResponse;
exports.toEcdsaSecp256r1ComputePublicKey = toEcdsaSecp256r1ComputePublicKey;
exports.fromEcdsaSecp256r1ComputePublicKey = fromEcdsaSecp256r1ComputePublicKey;
exports.toEcdsaSecp256r1ComputePublicKeyResponse = toEcdsaSecp256r1ComputePublicKeyResponse;
exports.fromEcdsaSecp256r1ComputePublicKeyResponse = fromEcdsaSecp256r1ComputePublicKeyResponse;
exports.toEcdsaSecp256r1ConstructSignature = toEcdsaSecp256r1ConstructSignature;
exports.fromEcdsaSecp256r1ConstructSignature = fromEcdsaSecp256r1ConstructSignature;
exports.toEcdsaSecp256r1ConstructSignatureResponse = toEcdsaSecp256r1ConstructSignatureResponse;
exports.fromEcdsaSecp256r1ConstructSignatureResponse = fromEcdsaSecp256r1ConstructSignatureResponse;
exports.toEcdsaSecp256r1RecoverPublicKey = toEcdsaSecp256r1RecoverPublicKey;
exports.fromEcdsaSecp256r1RecoverPublicKey = fromEcdsaSecp256r1RecoverPublicKey;
exports.toEcdsaSecp256r1RecoverPublicKeyResponse = toEcdsaSecp256r1RecoverPublicKeyResponse;
exports.fromEcdsaSecp256r1RecoverPublicKeyResponse = fromEcdsaSecp256r1RecoverPublicKeyResponse;
exports.toEcdsaSecp256r1VerifySignature = toEcdsaSecp256r1VerifySignature;
exports.fromEcdsaSecp256r1VerifySignature = fromEcdsaSecp256r1VerifySignature;
exports.toEcdsaSecp256r1VerifySignatureResponse = toEcdsaSecp256r1VerifySignatureResponse;
exports.fromEcdsaSecp256r1VerifySignatureResponse = fromEcdsaSecp256r1VerifySignatureResponse;
exports.toGoblinProof = toGoblinProof;
exports.fromGoblinProof = fromGoblinProof;
exports.toGrumpkinAdd = toGrumpkinAdd;
exports.fromGrumpkinAdd = fromGrumpkinAdd;
exports.toGrumpkinAddResponse = toGrumpkinAddResponse;
exports.fromGrumpkinAddResponse = fromGrumpkinAddResponse;
exports.toGrumpkinBatchMul = toGrumpkinBatchMul;
exports.fromGrumpkinBatchMul = fromGrumpkinBatchMul;
exports.toGrumpkinBatchMulResponse = toGrumpkinBatchMulResponse;
exports.fromGrumpkinBatchMulResponse = fromGrumpkinBatchMulResponse;
exports.toGrumpkinGetRandomFr = toGrumpkinGetRandomFr;
exports.fromGrumpkinGetRandomFr = fromGrumpkinGetRandomFr;
exports.toGrumpkinGetRandomFrResponse = toGrumpkinGetRandomFrResponse;
exports.fromGrumpkinGetRandomFrResponse = fromGrumpkinGetRandomFrResponse;
exports.toGrumpkinMul = toGrumpkinMul;
exports.fromGrumpkinMul = fromGrumpkinMul;
exports.toGrumpkinMulResponse = toGrumpkinMulResponse;
exports.fromGrumpkinMulResponse = fromGrumpkinMulResponse;
exports.toGrumpkinReduce512 = toGrumpkinReduce512;
exports.fromGrumpkinReduce512 = fromGrumpkinReduce512;
exports.toGrumpkinReduce512Response = toGrumpkinReduce512Response;
exports.fromGrumpkinReduce512Response = fromGrumpkinReduce512Response;
exports.toMegaVkAsFields = toMegaVkAsFields;
exports.fromMegaVkAsFields = fromMegaVkAsFields;
exports.toMegaVkAsFieldsResponse = toMegaVkAsFieldsResponse;
exports.fromMegaVkAsFieldsResponse = fromMegaVkAsFieldsResponse;
exports.toPedersenCommit = toPedersenCommit;
exports.fromPedersenCommit = fromPedersenCommit;
exports.toPedersenCommitResponse = toPedersenCommitResponse;
exports.fromPedersenCommitResponse = fromPedersenCommitResponse;
exports.toPedersenHash = toPedersenHash;
exports.fromPedersenHash = fromPedersenHash;
exports.toPedersenHashBuffer = toPedersenHashBuffer;
exports.fromPedersenHashBuffer = fromPedersenHashBuffer;
exports.toPedersenHashBufferResponse = toPedersenHashBufferResponse;
exports.fromPedersenHashBufferResponse = fromPedersenHashBufferResponse;
exports.toPedersenHashResponse = toPedersenHashResponse;
exports.fromPedersenHashResponse = fromPedersenHashResponse;
exports.toPoseidon2Hash = toPoseidon2Hash;
exports.fromPoseidon2Hash = fromPoseidon2Hash;
exports.toPoseidon2HashAccumulate = toPoseidon2HashAccumulate;
exports.fromPoseidon2HashAccumulate = fromPoseidon2HashAccumulate;
exports.toPoseidon2HashAccumulateResponse = toPoseidon2HashAccumulateResponse;
exports.fromPoseidon2HashAccumulateResponse = fromPoseidon2HashAccumulateResponse;
exports.toPoseidon2HashResponse = toPoseidon2HashResponse;
exports.fromPoseidon2HashResponse = fromPoseidon2HashResponse;
exports.toPoseidon2Permutation = toPoseidon2Permutation;
exports.fromPoseidon2Permutation = fromPoseidon2Permutation;
exports.toPoseidon2PermutationResponse = toPoseidon2PermutationResponse;
exports.fromPoseidon2PermutationResponse = fromPoseidon2PermutationResponse;
exports.toProofSystemSettings = toProofSystemSettings;
exports.fromProofSystemSettings = fromProofSystemSettings;
exports.toSchnorrComputePublicKey = toSchnorrComputePublicKey;
exports.fromSchnorrComputePublicKey = fromSchnorrComputePublicKey;
exports.toSchnorrComputePublicKeyResponse = toSchnorrComputePublicKeyResponse;
exports.fromSchnorrComputePublicKeyResponse = fromSchnorrComputePublicKeyResponse;
exports.toSchnorrConstructSignature = toSchnorrConstructSignature;
exports.fromSchnorrConstructSignature = fromSchnorrConstructSignature;
exports.toSchnorrConstructSignatureResponse = toSchnorrConstructSignatureResponse;
exports.fromSchnorrConstructSignatureResponse = fromSchnorrConstructSignatureResponse;
exports.toSchnorrVerifySignature = toSchnorrVerifySignature;
exports.fromSchnorrVerifySignature = fromSchnorrVerifySignature;
exports.toSchnorrVerifySignatureResponse = toSchnorrVerifySignatureResponse;
exports.fromSchnorrVerifySignatureResponse = fromSchnorrVerifySignatureResponse;
exports.toSecp256k1GetRandomFr = toSecp256k1GetRandomFr;
exports.fromSecp256k1GetRandomFr = fromSecp256k1GetRandomFr;
exports.toSecp256k1GetRandomFrResponse = toSecp256k1GetRandomFrResponse;
exports.fromSecp256k1GetRandomFrResponse = fromSecp256k1GetRandomFrResponse;
exports.toSecp256k1Mul = toSecp256k1Mul;
exports.fromSecp256k1Mul = fromSecp256k1Mul;
exports.toSecp256k1MulResponse = toSecp256k1MulResponse;
exports.fromSecp256k1MulResponse = fromSecp256k1MulResponse;
exports.toSecp256k1Reduce512 = toSecp256k1Reduce512;
exports.fromSecp256k1Reduce512 = fromSecp256k1Reduce512;
exports.toSecp256k1Reduce512Response = toSecp256k1Reduce512Response;
exports.fromSecp256k1Reduce512Response = fromSecp256k1Reduce512Response;
exports.toShutdown = toShutdown;
exports.fromShutdown = fromShutdown;
exports.toShutdownResponse = toShutdownResponse;
exports.fromShutdownResponse = fromShutdownResponse;
exports.toSrsInitGrumpkinSrs = toSrsInitGrumpkinSrs;
exports.fromSrsInitGrumpkinSrs = fromSrsInitGrumpkinSrs;
exports.toSrsInitGrumpkinSrsResponse = toSrsInitGrumpkinSrsResponse;
exports.fromSrsInitGrumpkinSrsResponse = fromSrsInitGrumpkinSrsResponse;
exports.toSrsInitSrs = toSrsInitSrs;
exports.fromSrsInitSrs = fromSrsInitSrs;
exports.toSrsInitSrsResponse = toSrsInitSrsResponse;
exports.fromSrsInitSrsResponse = fromSrsInitSrsResponse;
exports.toVkAsFields = toVkAsFields;
exports.fromVkAsFields = fromVkAsFields;
exports.toVkAsFieldsResponse = toVkAsFieldsResponse;
exports.fromVkAsFieldsResponse = fromVkAsFieldsResponse;
// Conversion functions (exported)
function toAesDecrypt(o) {
    if (o.ciphertext === undefined) {
        throw new Error("Expected ciphertext in AesDecrypt deserialization");
    }
    if (o.iv === undefined) {
        throw new Error("Expected iv in AesDecrypt deserialization");
    }
    if (o.key === undefined) {
        throw new Error("Expected key in AesDecrypt deserialization");
    }
    if (o.length === undefined) {
        throw new Error("Expected length in AesDecrypt deserialization");
    }
    ;
    return {
        ciphertext: o.ciphertext,
        iv: o.iv,
        key: o.key,
        length: o.length,
    };
}
function fromAesDecrypt(o) {
    if (o.ciphertext === undefined) {
        throw new Error("Expected ciphertext in AesDecrypt serialization");
    }
    if (o.iv === undefined) {
        throw new Error("Expected iv in AesDecrypt serialization");
    }
    if (o.key === undefined) {
        throw new Error("Expected key in AesDecrypt serialization");
    }
    if (o.length === undefined) {
        throw new Error("Expected length in AesDecrypt serialization");
    }
    ;
    return {
        ciphertext: o.ciphertext,
        iv: o.iv,
        key: o.key,
        length: o.length,
    };
}
function toAesDecryptResponse(o) {
    if (o.plaintext === undefined) {
        throw new Error("Expected plaintext in AesDecryptResponse deserialization");
    }
    ;
    return {
        plaintext: o.plaintext,
    };
}
function fromAesDecryptResponse(o) {
    if (o.plaintext === undefined) {
        throw new Error("Expected plaintext in AesDecryptResponse serialization");
    }
    ;
    return {
        plaintext: o.plaintext,
    };
}
function toAesEncrypt(o) {
    if (o.plaintext === undefined) {
        throw new Error("Expected plaintext in AesEncrypt deserialization");
    }
    if (o.iv === undefined) {
        throw new Error("Expected iv in AesEncrypt deserialization");
    }
    if (o.key === undefined) {
        throw new Error("Expected key in AesEncrypt deserialization");
    }
    if (o.length === undefined) {
        throw new Error("Expected length in AesEncrypt deserialization");
    }
    ;
    return {
        plaintext: o.plaintext,
        iv: o.iv,
        key: o.key,
        length: o.length,
    };
}
function fromAesEncrypt(o) {
    if (o.plaintext === undefined) {
        throw new Error("Expected plaintext in AesEncrypt serialization");
    }
    if (o.iv === undefined) {
        throw new Error("Expected iv in AesEncrypt serialization");
    }
    if (o.key === undefined) {
        throw new Error("Expected key in AesEncrypt serialization");
    }
    if (o.length === undefined) {
        throw new Error("Expected length in AesEncrypt serialization");
    }
    ;
    return {
        plaintext: o.plaintext,
        iv: o.iv,
        key: o.key,
        length: o.length,
    };
}
function toAesEncryptResponse(o) {
    if (o.ciphertext === undefined) {
        throw new Error("Expected ciphertext in AesEncryptResponse deserialization");
    }
    ;
    return {
        ciphertext: o.ciphertext,
    };
}
function fromAesEncryptResponse(o) {
    if (o.ciphertext === undefined) {
        throw new Error("Expected ciphertext in AesEncryptResponse serialization");
    }
    ;
    return {
        ciphertext: o.ciphertext,
    };
}
function toAffineElement(o) {
    if (o.x === undefined) {
        throw new Error("Expected x in AffineElement deserialization");
    }
    if (o.y === undefined) {
        throw new Error("Expected y in AffineElement deserialization");
    }
    ;
    return {
        x: o.x,
        y: o.y,
    };
}
function fromAffineElement(o) {
    if (o.x === undefined) {
        throw new Error("Expected x in AffineElement serialization");
    }
    if (o.y === undefined) {
        throw new Error("Expected y in AffineElement serialization");
    }
    ;
    return {
        x: o.x,
        y: o.y,
    };
}
function toBlake2s(o) {
    if (o.data === undefined) {
        throw new Error("Expected data in Blake2s deserialization");
    }
    ;
    return {
        data: o.data,
    };
}
function fromBlake2s(o) {
    if (o.data === undefined) {
        throw new Error("Expected data in Blake2s serialization");
    }
    ;
    return {
        data: o.data,
    };
}
function toBlake2sResponse(o) {
    if (o.hash === undefined) {
        throw new Error("Expected hash in Blake2sResponse deserialization");
    }
    ;
    return {
        hash: o.hash,
    };
}
function fromBlake2sResponse(o) {
    if (o.hash === undefined) {
        throw new Error("Expected hash in Blake2sResponse serialization");
    }
    ;
    return {
        hash: o.hash,
    };
}
function toBlake2sToField(o) {
    if (o.data === undefined) {
        throw new Error("Expected data in Blake2sToField deserialization");
    }
    ;
    return {
        data: o.data,
    };
}
function fromBlake2sToField(o) {
    if (o.data === undefined) {
        throw new Error("Expected data in Blake2sToField serialization");
    }
    ;
    return {
        data: o.data,
    };
}
function toBlake2sToFieldResponse(o) {
    if (o.field === undefined) {
        throw new Error("Expected field in Blake2sToFieldResponse deserialization");
    }
    ;
    return {
        field: o.field,
    };
}
function fromBlake2sToFieldResponse(o) {
    if (o.field === undefined) {
        throw new Error("Expected field in Blake2sToFieldResponse serialization");
    }
    ;
    return {
        field: o.field,
    };
}
function toBn254FrSqrt(o) {
    if (o.input === undefined) {
        throw new Error("Expected input in Bn254FrSqrt deserialization");
    }
    ;
    return {
        input: o.input,
    };
}
function fromBn254FrSqrt(o) {
    if (o.input === undefined) {
        throw new Error("Expected input in Bn254FrSqrt serialization");
    }
    ;
    return {
        input: o.input,
    };
}
function toBn254FrSqrtResponse(o) {
    if (o.is_square_root === undefined) {
        throw new Error("Expected is_square_root in Bn254FrSqrtResponse deserialization");
    }
    if (o.value === undefined) {
        throw new Error("Expected value in Bn254FrSqrtResponse deserialization");
    }
    ;
    return {
        isSquareRoot: o.is_square_root,
        value: o.value,
    };
}
function fromBn254FrSqrtResponse(o) {
    if (o.isSquareRoot === undefined) {
        throw new Error("Expected isSquareRoot in Bn254FrSqrtResponse serialization");
    }
    if (o.value === undefined) {
        throw new Error("Expected value in Bn254FrSqrtResponse serialization");
    }
    ;
    return {
        is_square_root: o.isSquareRoot,
        value: o.value,
    };
}
function toCircuitComputeVk(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in CircuitComputeVk deserialization");
    }
    if (o.settings === undefined) {
        throw new Error("Expected settings in CircuitComputeVk deserialization");
    }
    ;
    return {
        circuit: toCircuitInputNoVK(o.circuit),
        settings: toProofSystemSettings(o.settings),
    };
}
function fromCircuitComputeVk(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in CircuitComputeVk serialization");
    }
    if (o.settings === undefined) {
        throw new Error("Expected settings in CircuitComputeVk serialization");
    }
    ;
    return {
        circuit: fromCircuitInputNoVK(o.circuit),
        settings: fromProofSystemSettings(o.settings),
    };
}
function toCircuitComputeVkResponse(o) {
    if (o.bytes === undefined) {
        throw new Error("Expected bytes in CircuitComputeVkResponse deserialization");
    }
    if (o.fields === undefined) {
        throw new Error("Expected fields in CircuitComputeVkResponse deserialization");
    }
    if (o.hash === undefined) {
        throw new Error("Expected hash in CircuitComputeVkResponse deserialization");
    }
    ;
    return {
        bytes: o.bytes,
        fields: o.fields.map((v) => v),
        hash: o.hash,
    };
}
function fromCircuitComputeVkResponse(o) {
    if (o.bytes === undefined) {
        throw new Error("Expected bytes in CircuitComputeVkResponse serialization");
    }
    if (o.fields === undefined) {
        throw new Error("Expected fields in CircuitComputeVkResponse serialization");
    }
    if (o.hash === undefined) {
        throw new Error("Expected hash in CircuitComputeVkResponse serialization");
    }
    ;
    return {
        bytes: o.bytes,
        fields: o.fields.map((v) => v),
        hash: o.hash,
    };
}
function toCircuitInfoResponse(o) {
    if (o.num_gates === undefined) {
        throw new Error("Expected num_gates in CircuitInfoResponse deserialization");
    }
    if (o.num_gates_dyadic === undefined) {
        throw new Error("Expected num_gates_dyadic in CircuitInfoResponse deserialization");
    }
    if (o.num_acir_opcodes === undefined) {
        throw new Error("Expected num_acir_opcodes in CircuitInfoResponse deserialization");
    }
    if (o.gates_per_opcode === undefined) {
        throw new Error("Expected gates_per_opcode in CircuitInfoResponse deserialization");
    }
    ;
    return {
        numGates: o.num_gates,
        numGatesDyadic: o.num_gates_dyadic,
        numAcirOpcodes: o.num_acir_opcodes,
        gatesPerOpcode: o.gates_per_opcode.map((v) => v),
    };
}
function fromCircuitInfoResponse(o) {
    if (o.numGates === undefined) {
        throw new Error("Expected numGates in CircuitInfoResponse serialization");
    }
    if (o.numGatesDyadic === undefined) {
        throw new Error("Expected numGatesDyadic in CircuitInfoResponse serialization");
    }
    if (o.numAcirOpcodes === undefined) {
        throw new Error("Expected numAcirOpcodes in CircuitInfoResponse serialization");
    }
    if (o.gatesPerOpcode === undefined) {
        throw new Error("Expected gatesPerOpcode in CircuitInfoResponse serialization");
    }
    ;
    return {
        num_gates: o.numGates,
        num_gates_dyadic: o.numGatesDyadic,
        num_acir_opcodes: o.numAcirOpcodes,
        gates_per_opcode: o.gatesPerOpcode.map((v) => v),
    };
}
function toCircuitInput(o) {
    if (o.name === undefined) {
        throw new Error("Expected name in CircuitInput deserialization");
    }
    if (o.bytecode === undefined) {
        throw new Error("Expected bytecode in CircuitInput deserialization");
    }
    if (o.verification_key === undefined) {
        throw new Error("Expected verification_key in CircuitInput deserialization");
    }
    ;
    return {
        name: o.name,
        bytecode: o.bytecode,
        verificationKey: o.verification_key,
    };
}
function fromCircuitInput(o) {
    if (o.name === undefined) {
        throw new Error("Expected name in CircuitInput serialization");
    }
    if (o.bytecode === undefined) {
        throw new Error("Expected bytecode in CircuitInput serialization");
    }
    if (o.verificationKey === undefined) {
        throw new Error("Expected verificationKey in CircuitInput serialization");
    }
    ;
    return {
        name: o.name,
        bytecode: o.bytecode,
        verification_key: o.verificationKey,
    };
}
function toCircuitInputNoVK(o) {
    if (o.name === undefined) {
        throw new Error("Expected name in CircuitInputNoVK deserialization");
    }
    if (o.bytecode === undefined) {
        throw new Error("Expected bytecode in CircuitInputNoVK deserialization");
    }
    ;
    return {
        name: o.name,
        bytecode: o.bytecode,
    };
}
function fromCircuitInputNoVK(o) {
    if (o.name === undefined) {
        throw new Error("Expected name in CircuitInputNoVK serialization");
    }
    if (o.bytecode === undefined) {
        throw new Error("Expected bytecode in CircuitInputNoVK serialization");
    }
    ;
    return {
        name: o.name,
        bytecode: o.bytecode,
    };
}
function toCircuitProve(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in CircuitProve deserialization");
    }
    if (o.witness === undefined) {
        throw new Error("Expected witness in CircuitProve deserialization");
    }
    if (o.settings === undefined) {
        throw new Error("Expected settings in CircuitProve deserialization");
    }
    ;
    return {
        circuit: toCircuitInput(o.circuit),
        witness: o.witness,
        settings: toProofSystemSettings(o.settings),
    };
}
function fromCircuitProve(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in CircuitProve serialization");
    }
    if (o.witness === undefined) {
        throw new Error("Expected witness in CircuitProve serialization");
    }
    if (o.settings === undefined) {
        throw new Error("Expected settings in CircuitProve serialization");
    }
    ;
    return {
        circuit: fromCircuitInput(o.circuit),
        witness: o.witness,
        settings: fromProofSystemSettings(o.settings),
    };
}
function toCircuitProveResponse(o) {
    if (o.public_inputs === undefined) {
        throw new Error("Expected public_inputs in CircuitProveResponse deserialization");
    }
    if (o.proof === undefined) {
        throw new Error("Expected proof in CircuitProveResponse deserialization");
    }
    if (o.vk === undefined) {
        throw new Error("Expected vk in CircuitProveResponse deserialization");
    }
    ;
    return {
        publicInputs: o.public_inputs.map((v) => v),
        proof: o.proof.map((v) => v),
        vk: toCircuitComputeVkResponse(o.vk),
    };
}
function fromCircuitProveResponse(o) {
    if (o.publicInputs === undefined) {
        throw new Error("Expected publicInputs in CircuitProveResponse serialization");
    }
    if (o.proof === undefined) {
        throw new Error("Expected proof in CircuitProveResponse serialization");
    }
    if (o.vk === undefined) {
        throw new Error("Expected vk in CircuitProveResponse serialization");
    }
    ;
    return {
        public_inputs: o.publicInputs.map((v) => v),
        proof: o.proof.map((v) => v),
        vk: fromCircuitComputeVkResponse(o.vk),
    };
}
function toCircuitStats(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in CircuitStats deserialization");
    }
    if (o.include_gates_per_opcode === undefined) {
        throw new Error("Expected include_gates_per_opcode in CircuitStats deserialization");
    }
    if (o.settings === undefined) {
        throw new Error("Expected settings in CircuitStats deserialization");
    }
    ;
    return {
        circuit: toCircuitInput(o.circuit),
        includeGatesPerOpcode: o.include_gates_per_opcode,
        settings: toProofSystemSettings(o.settings),
    };
}
function fromCircuitStats(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in CircuitStats serialization");
    }
    if (o.includeGatesPerOpcode === undefined) {
        throw new Error("Expected includeGatesPerOpcode in CircuitStats serialization");
    }
    if (o.settings === undefined) {
        throw new Error("Expected settings in CircuitStats serialization");
    }
    ;
    return {
        circuit: fromCircuitInput(o.circuit),
        include_gates_per_opcode: o.includeGatesPerOpcode,
        settings: fromProofSystemSettings(o.settings),
    };
}
function toCircuitVerify(o) {
    if (o.verification_key === undefined) {
        throw new Error("Expected verification_key in CircuitVerify deserialization");
    }
    if (o.public_inputs === undefined) {
        throw new Error("Expected public_inputs in CircuitVerify deserialization");
    }
    if (o.proof === undefined) {
        throw new Error("Expected proof in CircuitVerify deserialization");
    }
    if (o.settings === undefined) {
        throw new Error("Expected settings in CircuitVerify deserialization");
    }
    ;
    return {
        verificationKey: o.verification_key,
        publicInputs: o.public_inputs.map((v) => v),
        proof: o.proof.map((v) => v),
        settings: toProofSystemSettings(o.settings),
    };
}
function fromCircuitVerify(o) {
    if (o.verificationKey === undefined) {
        throw new Error("Expected verificationKey in CircuitVerify serialization");
    }
    if (o.publicInputs === undefined) {
        throw new Error("Expected publicInputs in CircuitVerify serialization");
    }
    if (o.proof === undefined) {
        throw new Error("Expected proof in CircuitVerify serialization");
    }
    if (o.settings === undefined) {
        throw new Error("Expected settings in CircuitVerify serialization");
    }
    ;
    return {
        verification_key: o.verificationKey,
        public_inputs: o.publicInputs.map((v) => v),
        proof: o.proof.map((v) => v),
        settings: fromProofSystemSettings(o.settings),
    };
}
function toCircuitVerifyResponse(o) {
    if (o.verified === undefined) {
        throw new Error("Expected verified in CircuitVerifyResponse deserialization");
    }
    ;
    return {
        verified: o.verified,
    };
}
function fromCircuitVerifyResponse(o) {
    if (o.verified === undefined) {
        throw new Error("Expected verified in CircuitVerifyResponse serialization");
    }
    ;
    return {
        verified: o.verified,
    };
}
function toCircuitWriteSolidityVerifier(o) {
    if (o.verification_key === undefined) {
        throw new Error("Expected verification_key in CircuitWriteSolidityVerifier deserialization");
    }
    if (o.settings === undefined) {
        throw new Error("Expected settings in CircuitWriteSolidityVerifier deserialization");
    }
    ;
    return {
        verificationKey: o.verification_key,
        settings: toProofSystemSettings(o.settings),
    };
}
function fromCircuitWriteSolidityVerifier(o) {
    if (o.verificationKey === undefined) {
        throw new Error("Expected verificationKey in CircuitWriteSolidityVerifier serialization");
    }
    if (o.settings === undefined) {
        throw new Error("Expected settings in CircuitWriteSolidityVerifier serialization");
    }
    ;
    return {
        verification_key: o.verificationKey,
        settings: fromProofSystemSettings(o.settings),
    };
}
function toCircuitWriteSolidityVerifierResponse(o) {
    if (o.solidity_code === undefined) {
        throw new Error("Expected solidity_code in CircuitWriteSolidityVerifierResponse deserialization");
    }
    ;
    return {
        solidityCode: o.solidity_code,
    };
}
function fromCircuitWriteSolidityVerifierResponse(o) {
    if (o.solidityCode === undefined) {
        throw new Error("Expected solidityCode in CircuitWriteSolidityVerifierResponse serialization");
    }
    ;
    return {
        solidity_code: o.solidityCode,
    };
}
function toClientIvcAccumulate(o) {
    if (o.witness === undefined) {
        throw new Error("Expected witness in ClientIvcAccumulate deserialization");
    }
    ;
    return {
        witness: o.witness,
    };
}
function fromClientIvcAccumulate(o) {
    if (o.witness === undefined) {
        throw new Error("Expected witness in ClientIvcAccumulate serialization");
    }
    ;
    return {
        witness: o.witness,
    };
}
function toClientIvcAccumulateResponse(o) {
    return {};
}
function fromClientIvcAccumulateResponse(o) {
    return {};
}
function toClientIvcCheckPrecomputedVk(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in ClientIvcCheckPrecomputedVk deserialization");
    }
    ;
    return {
        circuit: toCircuitInput(o.circuit),
    };
}
function fromClientIvcCheckPrecomputedVk(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in ClientIvcCheckPrecomputedVk serialization");
    }
    ;
    return {
        circuit: fromCircuitInput(o.circuit),
    };
}
function toClientIvcCheckPrecomputedVkResponse(o) {
    if (o.valid === undefined) {
        throw new Error("Expected valid in ClientIvcCheckPrecomputedVkResponse deserialization");
    }
    if (o.actual_vk === undefined) {
        throw new Error("Expected actual_vk in ClientIvcCheckPrecomputedVkResponse deserialization");
    }
    ;
    return {
        valid: o.valid,
        actualVk: o.actual_vk,
    };
}
function fromClientIvcCheckPrecomputedVkResponse(o) {
    if (o.valid === undefined) {
        throw new Error("Expected valid in ClientIvcCheckPrecomputedVkResponse serialization");
    }
    if (o.actualVk === undefined) {
        throw new Error("Expected actualVk in ClientIvcCheckPrecomputedVkResponse serialization");
    }
    ;
    return {
        valid: o.valid,
        actual_vk: o.actualVk,
    };
}
function toClientIvcComputeIvcVk(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in ClientIvcComputeIvcVk deserialization");
    }
    ;
    return {
        circuit: toCircuitInputNoVK(o.circuit),
    };
}
function fromClientIvcComputeIvcVk(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in ClientIvcComputeIvcVk serialization");
    }
    ;
    return {
        circuit: fromCircuitInputNoVK(o.circuit),
    };
}
function toClientIvcComputeIvcVkResponse(o) {
    if (o.bytes === undefined) {
        throw new Error("Expected bytes in ClientIvcComputeIvcVkResponse deserialization");
    }
    ;
    return {
        bytes: o.bytes,
    };
}
function fromClientIvcComputeIvcVkResponse(o) {
    if (o.bytes === undefined) {
        throw new Error("Expected bytes in ClientIvcComputeIvcVkResponse serialization");
    }
    ;
    return {
        bytes: o.bytes,
    };
}
function toClientIvcComputeStandaloneVk(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in ClientIvcComputeStandaloneVk deserialization");
    }
    ;
    return {
        circuit: toCircuitInputNoVK(o.circuit),
    };
}
function fromClientIvcComputeStandaloneVk(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in ClientIvcComputeStandaloneVk serialization");
    }
    ;
    return {
        circuit: fromCircuitInputNoVK(o.circuit),
    };
}
function toClientIvcComputeStandaloneVkResponse(o) {
    if (o.bytes === undefined) {
        throw new Error("Expected bytes in ClientIvcComputeStandaloneVkResponse deserialization");
    }
    if (o.fields === undefined) {
        throw new Error("Expected fields in ClientIvcComputeStandaloneVkResponse deserialization");
    }
    ;
    return {
        bytes: o.bytes,
        fields: o.fields.map((v) => v),
    };
}
function fromClientIvcComputeStandaloneVkResponse(o) {
    if (o.bytes === undefined) {
        throw new Error("Expected bytes in ClientIvcComputeStandaloneVkResponse serialization");
    }
    if (o.fields === undefined) {
        throw new Error("Expected fields in ClientIvcComputeStandaloneVkResponse serialization");
    }
    ;
    return {
        bytes: o.bytes,
        fields: o.fields.map((v) => v),
    };
}
function toClientIvcLoad(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in ClientIvcLoad deserialization");
    }
    ;
    return {
        circuit: toCircuitInput(o.circuit),
    };
}
function fromClientIvcLoad(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in ClientIvcLoad serialization");
    }
    ;
    return {
        circuit: fromCircuitInput(o.circuit),
    };
}
function toClientIvcLoadResponse(o) {
    return {};
}
function fromClientIvcLoadResponse(o) {
    return {};
}
function toClientIVCProof(o) {
    if (o.mega_proof === undefined) {
        throw new Error("Expected mega_proof in ClientIVCProof deserialization");
    }
    if (o.goblin_proof === undefined) {
        throw new Error("Expected goblin_proof in ClientIVCProof deserialization");
    }
    ;
    return {
        megaProof: o.mega_proof.map((v) => v),
        goblinProof: toGoblinProof(o.goblin_proof),
    };
}
function fromClientIVCProof(o) {
    if (o.megaProof === undefined) {
        throw new Error("Expected megaProof in ClientIVCProof serialization");
    }
    if (o.goblinProof === undefined) {
        throw new Error("Expected goblinProof in ClientIVCProof serialization");
    }
    ;
    return {
        mega_proof: o.megaProof.map((v) => v),
        goblin_proof: fromGoblinProof(o.goblinProof),
    };
}
function toClientIvcProve(o) {
    return {};
}
function fromClientIvcProve(o) {
    return {};
}
function toClientIvcProveResponse(o) {
    if (o.proof === undefined) {
        throw new Error("Expected proof in ClientIvcProveResponse deserialization");
    }
    ;
    return {
        proof: toClientIVCProof(o.proof),
    };
}
function fromClientIvcProveResponse(o) {
    if (o.proof === undefined) {
        throw new Error("Expected proof in ClientIvcProveResponse serialization");
    }
    ;
    return {
        proof: fromClientIVCProof(o.proof),
    };
}
function toClientIvcStart(o) {
    if (o.num_circuits === undefined) {
        throw new Error("Expected num_circuits in ClientIvcStart deserialization");
    }
    ;
    return {
        numCircuits: o.num_circuits,
    };
}
function fromClientIvcStart(o) {
    if (o.numCircuits === undefined) {
        throw new Error("Expected numCircuits in ClientIvcStart serialization");
    }
    ;
    return {
        num_circuits: o.numCircuits,
    };
}
function toClientIvcStartResponse(o) {
    return {};
}
function fromClientIvcStartResponse(o) {
    return {};
}
function toClientIvcStats(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in ClientIvcStats deserialization");
    }
    if (o.include_gates_per_opcode === undefined) {
        throw new Error("Expected include_gates_per_opcode in ClientIvcStats deserialization");
    }
    ;
    return {
        circuit: toCircuitInputNoVK(o.circuit),
        includeGatesPerOpcode: o.include_gates_per_opcode,
    };
}
function fromClientIvcStats(o) {
    if (o.circuit === undefined) {
        throw new Error("Expected circuit in ClientIvcStats serialization");
    }
    if (o.includeGatesPerOpcode === undefined) {
        throw new Error("Expected includeGatesPerOpcode in ClientIvcStats serialization");
    }
    ;
    return {
        circuit: fromCircuitInputNoVK(o.circuit),
        include_gates_per_opcode: o.includeGatesPerOpcode,
    };
}
function toClientIvcStatsResponse(o) {
    if (o.acir_opcodes === undefined) {
        throw new Error("Expected acir_opcodes in ClientIvcStatsResponse deserialization");
    }
    if (o.circuit_size === undefined) {
        throw new Error("Expected circuit_size in ClientIvcStatsResponse deserialization");
    }
    if (o.gates_per_opcode === undefined) {
        throw new Error("Expected gates_per_opcode in ClientIvcStatsResponse deserialization");
    }
    ;
    return {
        acirOpcodes: o.acir_opcodes,
        circuitSize: o.circuit_size,
        gatesPerOpcode: o.gates_per_opcode.map((v) => v),
    };
}
function fromClientIvcStatsResponse(o) {
    if (o.acirOpcodes === undefined) {
        throw new Error("Expected acirOpcodes in ClientIvcStatsResponse serialization");
    }
    if (o.circuitSize === undefined) {
        throw new Error("Expected circuitSize in ClientIvcStatsResponse serialization");
    }
    if (o.gatesPerOpcode === undefined) {
        throw new Error("Expected gatesPerOpcode in ClientIvcStatsResponse serialization");
    }
    ;
    return {
        acir_opcodes: o.acirOpcodes,
        circuit_size: o.circuitSize,
        gates_per_opcode: o.gatesPerOpcode.map((v) => v),
    };
}
function toClientIvcVerify(o) {
    if (o.proof === undefined) {
        throw new Error("Expected proof in ClientIvcVerify deserialization");
    }
    if (o.vk === undefined) {
        throw new Error("Expected vk in ClientIvcVerify deserialization");
    }
    ;
    return {
        proof: toClientIVCProof(o.proof),
        vk: o.vk,
    };
}
function fromClientIvcVerify(o) {
    if (o.proof === undefined) {
        throw new Error("Expected proof in ClientIvcVerify serialization");
    }
    if (o.vk === undefined) {
        throw new Error("Expected vk in ClientIvcVerify serialization");
    }
    ;
    return {
        proof: fromClientIVCProof(o.proof),
        vk: o.vk,
    };
}
function toClientIvcVerifyResponse(o) {
    if (o.valid === undefined) {
        throw new Error("Expected valid in ClientIvcVerifyResponse deserialization");
    }
    ;
    return {
        valid: o.valid,
    };
}
function fromClientIvcVerifyResponse(o) {
    if (o.valid === undefined) {
        throw new Error("Expected valid in ClientIvcVerifyResponse serialization");
    }
    ;
    return {
        valid: o.valid,
    };
}
function toECCVMProof(o) {
    if (o.pre_ipa_proof === undefined) {
        throw new Error("Expected pre_ipa_proof in ECCVMProof deserialization");
    }
    if (o.ipa_proof === undefined) {
        throw new Error("Expected ipa_proof in ECCVMProof deserialization");
    }
    ;
    return {
        preIpaProof: o.pre_ipa_proof.map((v) => v),
        ipaProof: o.ipa_proof.map((v) => v),
    };
}
function fromECCVMProof(o) {
    if (o.preIpaProof === undefined) {
        throw new Error("Expected preIpaProof in ECCVMProof serialization");
    }
    if (o.ipaProof === undefined) {
        throw new Error("Expected ipaProof in ECCVMProof serialization");
    }
    ;
    return {
        pre_ipa_proof: o.preIpaProof.map((v) => v),
        ipa_proof: o.ipaProof.map((v) => v),
    };
}
function toEcdsaSecp256k1ComputePublicKey(o) {
    if (o.private_key === undefined) {
        throw new Error("Expected private_key in EcdsaSecp256k1ComputePublicKey deserialization");
    }
    ;
    return {
        privateKey: o.private_key,
    };
}
function fromEcdsaSecp256k1ComputePublicKey(o) {
    if (o.privateKey === undefined) {
        throw new Error("Expected privateKey in EcdsaSecp256k1ComputePublicKey serialization");
    }
    ;
    return {
        private_key: o.privateKey,
    };
}
function toEcdsaSecp256k1ComputePublicKeyResponse(o) {
    if (o.public_key === undefined) {
        throw new Error("Expected public_key in EcdsaSecp256k1ComputePublicKeyResponse deserialization");
    }
    ;
    return {
        publicKey: toAffineElement(o.public_key),
    };
}
function fromEcdsaSecp256k1ComputePublicKeyResponse(o) {
    if (o.publicKey === undefined) {
        throw new Error("Expected publicKey in EcdsaSecp256k1ComputePublicKeyResponse serialization");
    }
    ;
    return {
        public_key: fromAffineElement(o.publicKey),
    };
}
function toEcdsaSecp256k1ConstructSignature(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256k1ConstructSignature deserialization");
    }
    if (o.private_key === undefined) {
        throw new Error("Expected private_key in EcdsaSecp256k1ConstructSignature deserialization");
    }
    ;
    return {
        message: o.message,
        privateKey: o.private_key,
    };
}
function fromEcdsaSecp256k1ConstructSignature(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256k1ConstructSignature serialization");
    }
    if (o.privateKey === undefined) {
        throw new Error("Expected privateKey in EcdsaSecp256k1ConstructSignature serialization");
    }
    ;
    return {
        message: o.message,
        private_key: o.privateKey,
    };
}
function toEcdsaSecp256k1ConstructSignatureResponse(o) {
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256k1ConstructSignatureResponse deserialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256k1ConstructSignatureResponse deserialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256k1ConstructSignatureResponse deserialization");
    }
    ;
    return {
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
function fromEcdsaSecp256k1ConstructSignatureResponse(o) {
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256k1ConstructSignatureResponse serialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256k1ConstructSignatureResponse serialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256k1ConstructSignatureResponse serialization");
    }
    ;
    return {
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
function toEcdsaSecp256k1RecoverPublicKey(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256k1RecoverPublicKey deserialization");
    }
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256k1RecoverPublicKey deserialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256k1RecoverPublicKey deserialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256k1RecoverPublicKey deserialization");
    }
    ;
    return {
        message: o.message,
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
function fromEcdsaSecp256k1RecoverPublicKey(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256k1RecoverPublicKey serialization");
    }
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256k1RecoverPublicKey serialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256k1RecoverPublicKey serialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256k1RecoverPublicKey serialization");
    }
    ;
    return {
        message: o.message,
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
function toEcdsaSecp256k1RecoverPublicKeyResponse(o) {
    if (o.public_key === undefined) {
        throw new Error("Expected public_key in EcdsaSecp256k1RecoverPublicKeyResponse deserialization");
    }
    ;
    return {
        publicKey: toAffineElement(o.public_key),
    };
}
function fromEcdsaSecp256k1RecoverPublicKeyResponse(o) {
    if (o.publicKey === undefined) {
        throw new Error("Expected publicKey in EcdsaSecp256k1RecoverPublicKeyResponse serialization");
    }
    ;
    return {
        public_key: fromAffineElement(o.publicKey),
    };
}
function toEcdsaSecp256k1VerifySignature(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256k1VerifySignature deserialization");
    }
    if (o.public_key === undefined) {
        throw new Error("Expected public_key in EcdsaSecp256k1VerifySignature deserialization");
    }
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256k1VerifySignature deserialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256k1VerifySignature deserialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256k1VerifySignature deserialization");
    }
    ;
    return {
        message: o.message,
        publicKey: toAffineElement(o.public_key),
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
function fromEcdsaSecp256k1VerifySignature(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256k1VerifySignature serialization");
    }
    if (o.publicKey === undefined) {
        throw new Error("Expected publicKey in EcdsaSecp256k1VerifySignature serialization");
    }
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256k1VerifySignature serialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256k1VerifySignature serialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256k1VerifySignature serialization");
    }
    ;
    return {
        message: o.message,
        public_key: fromAffineElement(o.publicKey),
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
function toEcdsaSecp256k1VerifySignatureResponse(o) {
    if (o.verified === undefined) {
        throw new Error("Expected verified in EcdsaSecp256k1VerifySignatureResponse deserialization");
    }
    ;
    return {
        verified: o.verified,
    };
}
function fromEcdsaSecp256k1VerifySignatureResponse(o) {
    if (o.verified === undefined) {
        throw new Error("Expected verified in EcdsaSecp256k1VerifySignatureResponse serialization");
    }
    ;
    return {
        verified: o.verified,
    };
}
function toEcdsaSecp256r1ComputePublicKey(o) {
    if (o.private_key === undefined) {
        throw new Error("Expected private_key in EcdsaSecp256r1ComputePublicKey deserialization");
    }
    ;
    return {
        privateKey: o.private_key,
    };
}
function fromEcdsaSecp256r1ComputePublicKey(o) {
    if (o.privateKey === undefined) {
        throw new Error("Expected privateKey in EcdsaSecp256r1ComputePublicKey serialization");
    }
    ;
    return {
        private_key: o.privateKey,
    };
}
function toEcdsaSecp256r1ComputePublicKeyResponse(o) {
    if (o.public_key === undefined) {
        throw new Error("Expected public_key in EcdsaSecp256r1ComputePublicKeyResponse deserialization");
    }
    ;
    return {
        publicKey: toAffineElement(o.public_key),
    };
}
function fromEcdsaSecp256r1ComputePublicKeyResponse(o) {
    if (o.publicKey === undefined) {
        throw new Error("Expected publicKey in EcdsaSecp256r1ComputePublicKeyResponse serialization");
    }
    ;
    return {
        public_key: fromAffineElement(o.publicKey),
    };
}
function toEcdsaSecp256r1ConstructSignature(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256r1ConstructSignature deserialization");
    }
    if (o.private_key === undefined) {
        throw new Error("Expected private_key in EcdsaSecp256r1ConstructSignature deserialization");
    }
    ;
    return {
        message: o.message,
        privateKey: o.private_key,
    };
}
function fromEcdsaSecp256r1ConstructSignature(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256r1ConstructSignature serialization");
    }
    if (o.privateKey === undefined) {
        throw new Error("Expected privateKey in EcdsaSecp256r1ConstructSignature serialization");
    }
    ;
    return {
        message: o.message,
        private_key: o.privateKey,
    };
}
function toEcdsaSecp256r1ConstructSignatureResponse(o) {
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256r1ConstructSignatureResponse deserialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256r1ConstructSignatureResponse deserialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256r1ConstructSignatureResponse deserialization");
    }
    ;
    return {
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
function fromEcdsaSecp256r1ConstructSignatureResponse(o) {
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256r1ConstructSignatureResponse serialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256r1ConstructSignatureResponse serialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256r1ConstructSignatureResponse serialization");
    }
    ;
    return {
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
function toEcdsaSecp256r1RecoverPublicKey(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256r1RecoverPublicKey deserialization");
    }
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256r1RecoverPublicKey deserialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256r1RecoverPublicKey deserialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256r1RecoverPublicKey deserialization");
    }
    ;
    return {
        message: o.message,
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
function fromEcdsaSecp256r1RecoverPublicKey(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256r1RecoverPublicKey serialization");
    }
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256r1RecoverPublicKey serialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256r1RecoverPublicKey serialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256r1RecoverPublicKey serialization");
    }
    ;
    return {
        message: o.message,
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
function toEcdsaSecp256r1RecoverPublicKeyResponse(o) {
    if (o.public_key === undefined) {
        throw new Error("Expected public_key in EcdsaSecp256r1RecoverPublicKeyResponse deserialization");
    }
    ;
    return {
        publicKey: toAffineElement(o.public_key),
    };
}
function fromEcdsaSecp256r1RecoverPublicKeyResponse(o) {
    if (o.publicKey === undefined) {
        throw new Error("Expected publicKey in EcdsaSecp256r1RecoverPublicKeyResponse serialization");
    }
    ;
    return {
        public_key: fromAffineElement(o.publicKey),
    };
}
function toEcdsaSecp256r1VerifySignature(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256r1VerifySignature deserialization");
    }
    if (o.public_key === undefined) {
        throw new Error("Expected public_key in EcdsaSecp256r1VerifySignature deserialization");
    }
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256r1VerifySignature deserialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256r1VerifySignature deserialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256r1VerifySignature deserialization");
    }
    ;
    return {
        message: o.message,
        publicKey: toAffineElement(o.public_key),
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
function fromEcdsaSecp256r1VerifySignature(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in EcdsaSecp256r1VerifySignature serialization");
    }
    if (o.publicKey === undefined) {
        throw new Error("Expected publicKey in EcdsaSecp256r1VerifySignature serialization");
    }
    if (o.r === undefined) {
        throw new Error("Expected r in EcdsaSecp256r1VerifySignature serialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in EcdsaSecp256r1VerifySignature serialization");
    }
    if (o.v === undefined) {
        throw new Error("Expected v in EcdsaSecp256r1VerifySignature serialization");
    }
    ;
    return {
        message: o.message,
        public_key: fromAffineElement(o.publicKey),
        r: o.r,
        s: o.s,
        v: o.v,
    };
}
function toEcdsaSecp256r1VerifySignatureResponse(o) {
    if (o.verified === undefined) {
        throw new Error("Expected verified in EcdsaSecp256r1VerifySignatureResponse deserialization");
    }
    ;
    return {
        verified: o.verified,
    };
}
function fromEcdsaSecp256r1VerifySignatureResponse(o) {
    if (o.verified === undefined) {
        throw new Error("Expected verified in EcdsaSecp256r1VerifySignatureResponse serialization");
    }
    ;
    return {
        verified: o.verified,
    };
}
function toGoblinProof(o) {
    if (o.merge_proof === undefined) {
        throw new Error("Expected merge_proof in GoblinProof deserialization");
    }
    if (o.eccvm_proof === undefined) {
        throw new Error("Expected eccvm_proof in GoblinProof deserialization");
    }
    if (o.translator_proof === undefined) {
        throw new Error("Expected translator_proof in GoblinProof deserialization");
    }
    ;
    return {
        mergeProof: o.merge_proof.map((v) => v),
        eccvmProof: toECCVMProof(o.eccvm_proof),
        translatorProof: o.translator_proof.map((v) => v),
    };
}
function fromGoblinProof(o) {
    if (o.mergeProof === undefined) {
        throw new Error("Expected mergeProof in GoblinProof serialization");
    }
    if (o.eccvmProof === undefined) {
        throw new Error("Expected eccvmProof in GoblinProof serialization");
    }
    if (o.translatorProof === undefined) {
        throw new Error("Expected translatorProof in GoblinProof serialization");
    }
    ;
    return {
        merge_proof: o.mergeProof.map((v) => v),
        eccvm_proof: fromECCVMProof(o.eccvmProof),
        translator_proof: o.translatorProof.map((v) => v),
    };
}
function toGrumpkinAdd(o) {
    if (o.point_a === undefined) {
        throw new Error("Expected point_a in GrumpkinAdd deserialization");
    }
    if (o.point_b === undefined) {
        throw new Error("Expected point_b in GrumpkinAdd deserialization");
    }
    ;
    return {
        pointA: toAffineElement(o.point_a),
        pointB: toAffineElement(o.point_b),
    };
}
function fromGrumpkinAdd(o) {
    if (o.pointA === undefined) {
        throw new Error("Expected pointA in GrumpkinAdd serialization");
    }
    if (o.pointB === undefined) {
        throw new Error("Expected pointB in GrumpkinAdd serialization");
    }
    ;
    return {
        point_a: fromAffineElement(o.pointA),
        point_b: fromAffineElement(o.pointB),
    };
}
function toGrumpkinAddResponse(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in GrumpkinAddResponse deserialization");
    }
    ;
    return {
        point: toAffineElement(o.point),
    };
}
function fromGrumpkinAddResponse(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in GrumpkinAddResponse serialization");
    }
    ;
    return {
        point: fromAffineElement(o.point),
    };
}
function toGrumpkinBatchMul(o) {
    if (o.points === undefined) {
        throw new Error("Expected points in GrumpkinBatchMul deserialization");
    }
    if (o.scalar === undefined) {
        throw new Error("Expected scalar in GrumpkinBatchMul deserialization");
    }
    ;
    return {
        points: o.points.map((v) => v),
        scalar: o.scalar,
    };
}
function fromGrumpkinBatchMul(o) {
    if (o.points === undefined) {
        throw new Error("Expected points in GrumpkinBatchMul serialization");
    }
    if (o.scalar === undefined) {
        throw new Error("Expected scalar in GrumpkinBatchMul serialization");
    }
    ;
    return {
        points: o.points.map((v) => v),
        scalar: o.scalar,
    };
}
function toGrumpkinBatchMulResponse(o) {
    if (o.points === undefined) {
        throw new Error("Expected points in GrumpkinBatchMulResponse deserialization");
    }
    ;
    return {
        points: o.points.map((v) => v),
    };
}
function fromGrumpkinBatchMulResponse(o) {
    if (o.points === undefined) {
        throw new Error("Expected points in GrumpkinBatchMulResponse serialization");
    }
    ;
    return {
        points: o.points.map((v) => v),
    };
}
function toGrumpkinGetRandomFr(o) {
    if (o.dummy === undefined) {
        throw new Error("Expected dummy in GrumpkinGetRandomFr deserialization");
    }
    ;
    return {
        dummy: o.dummy,
    };
}
function fromGrumpkinGetRandomFr(o) {
    if (o.dummy === undefined) {
        throw new Error("Expected dummy in GrumpkinGetRandomFr serialization");
    }
    ;
    return {
        dummy: o.dummy,
    };
}
function toGrumpkinGetRandomFrResponse(o) {
    if (o.value === undefined) {
        throw new Error("Expected value in GrumpkinGetRandomFrResponse deserialization");
    }
    ;
    return {
        value: o.value,
    };
}
function fromGrumpkinGetRandomFrResponse(o) {
    if (o.value === undefined) {
        throw new Error("Expected value in GrumpkinGetRandomFrResponse serialization");
    }
    ;
    return {
        value: o.value,
    };
}
function toGrumpkinMul(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in GrumpkinMul deserialization");
    }
    if (o.scalar === undefined) {
        throw new Error("Expected scalar in GrumpkinMul deserialization");
    }
    ;
    return {
        point: toAffineElement(o.point),
        scalar: o.scalar,
    };
}
function fromGrumpkinMul(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in GrumpkinMul serialization");
    }
    if (o.scalar === undefined) {
        throw new Error("Expected scalar in GrumpkinMul serialization");
    }
    ;
    return {
        point: fromAffineElement(o.point),
        scalar: o.scalar,
    };
}
function toGrumpkinMulResponse(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in GrumpkinMulResponse deserialization");
    }
    ;
    return {
        point: toAffineElement(o.point),
    };
}
function fromGrumpkinMulResponse(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in GrumpkinMulResponse serialization");
    }
    ;
    return {
        point: fromAffineElement(o.point),
    };
}
function toGrumpkinReduce512(o) {
    if (o.input === undefined) {
        throw new Error("Expected input in GrumpkinReduce512 deserialization");
    }
    ;
    return {
        input: o.input,
    };
}
function fromGrumpkinReduce512(o) {
    if (o.input === undefined) {
        throw new Error("Expected input in GrumpkinReduce512 serialization");
    }
    ;
    return {
        input: o.input,
    };
}
function toGrumpkinReduce512Response(o) {
    if (o.value === undefined) {
        throw new Error("Expected value in GrumpkinReduce512Response deserialization");
    }
    ;
    return {
        value: o.value,
    };
}
function fromGrumpkinReduce512Response(o) {
    if (o.value === undefined) {
        throw new Error("Expected value in GrumpkinReduce512Response serialization");
    }
    ;
    return {
        value: o.value,
    };
}
function toMegaVkAsFields(o) {
    if (o.verification_key === undefined) {
        throw new Error("Expected verification_key in MegaVkAsFields deserialization");
    }
    ;
    return {
        verificationKey: o.verification_key,
    };
}
function fromMegaVkAsFields(o) {
    if (o.verificationKey === undefined) {
        throw new Error("Expected verificationKey in MegaVkAsFields serialization");
    }
    ;
    return {
        verification_key: o.verificationKey,
    };
}
function toMegaVkAsFieldsResponse(o) {
    if (o.fields === undefined) {
        throw new Error("Expected fields in MegaVkAsFieldsResponse deserialization");
    }
    ;
    return {
        fields: o.fields.map((v) => v),
    };
}
function fromMegaVkAsFieldsResponse(o) {
    if (o.fields === undefined) {
        throw new Error("Expected fields in MegaVkAsFieldsResponse serialization");
    }
    ;
    return {
        fields: o.fields.map((v) => v),
    };
}
function toPedersenCommit(o) {
    if (o.inputs === undefined) {
        throw new Error("Expected inputs in PedersenCommit deserialization");
    }
    if (o.hash_index === undefined) {
        throw new Error("Expected hash_index in PedersenCommit deserialization");
    }
    ;
    return {
        inputs: o.inputs.map((v) => v),
        hashIndex: o.hash_index,
    };
}
function fromPedersenCommit(o) {
    if (o.inputs === undefined) {
        throw new Error("Expected inputs in PedersenCommit serialization");
    }
    if (o.hashIndex === undefined) {
        throw new Error("Expected hashIndex in PedersenCommit serialization");
    }
    ;
    return {
        inputs: o.inputs.map((v) => v),
        hash_index: o.hashIndex,
    };
}
function toPedersenCommitResponse(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in PedersenCommitResponse deserialization");
    }
    ;
    return {
        point: toAffineElement(o.point),
    };
}
function fromPedersenCommitResponse(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in PedersenCommitResponse serialization");
    }
    ;
    return {
        point: fromAffineElement(o.point),
    };
}
function toPedersenHash(o) {
    if (o.inputs === undefined) {
        throw new Error("Expected inputs in PedersenHash deserialization");
    }
    if (o.hash_index === undefined) {
        throw new Error("Expected hash_index in PedersenHash deserialization");
    }
    ;
    return {
        inputs: o.inputs.map((v) => v),
        hashIndex: o.hash_index,
    };
}
function fromPedersenHash(o) {
    if (o.inputs === undefined) {
        throw new Error("Expected inputs in PedersenHash serialization");
    }
    if (o.hashIndex === undefined) {
        throw new Error("Expected hashIndex in PedersenHash serialization");
    }
    ;
    return {
        inputs: o.inputs.map((v) => v),
        hash_index: o.hashIndex,
    };
}
function toPedersenHashBuffer(o) {
    if (o.input === undefined) {
        throw new Error("Expected input in PedersenHashBuffer deserialization");
    }
    if (o.hash_index === undefined) {
        throw new Error("Expected hash_index in PedersenHashBuffer deserialization");
    }
    ;
    return {
        input: o.input,
        hashIndex: o.hash_index,
    };
}
function fromPedersenHashBuffer(o) {
    if (o.input === undefined) {
        throw new Error("Expected input in PedersenHashBuffer serialization");
    }
    if (o.hashIndex === undefined) {
        throw new Error("Expected hashIndex in PedersenHashBuffer serialization");
    }
    ;
    return {
        input: o.input,
        hash_index: o.hashIndex,
    };
}
function toPedersenHashBufferResponse(o) {
    if (o.hash === undefined) {
        throw new Error("Expected hash in PedersenHashBufferResponse deserialization");
    }
    ;
    return {
        hash: o.hash,
    };
}
function fromPedersenHashBufferResponse(o) {
    if (o.hash === undefined) {
        throw new Error("Expected hash in PedersenHashBufferResponse serialization");
    }
    ;
    return {
        hash: o.hash,
    };
}
function toPedersenHashResponse(o) {
    if (o.hash === undefined) {
        throw new Error("Expected hash in PedersenHashResponse deserialization");
    }
    ;
    return {
        hash: o.hash,
    };
}
function fromPedersenHashResponse(o) {
    if (o.hash === undefined) {
        throw new Error("Expected hash in PedersenHashResponse serialization");
    }
    ;
    return {
        hash: o.hash,
    };
}
function toPoseidon2Hash(o) {
    if (o.inputs === undefined) {
        throw new Error("Expected inputs in Poseidon2Hash deserialization");
    }
    ;
    return {
        inputs: o.inputs.map((v) => v),
    };
}
function fromPoseidon2Hash(o) {
    if (o.inputs === undefined) {
        throw new Error("Expected inputs in Poseidon2Hash serialization");
    }
    ;
    return {
        inputs: o.inputs.map((v) => v),
    };
}
function toPoseidon2HashAccumulate(o) {
    if (o.inputs === undefined) {
        throw new Error("Expected inputs in Poseidon2HashAccumulate deserialization");
    }
    ;
    return {
        inputs: o.inputs.map((v) => v),
    };
}
function fromPoseidon2HashAccumulate(o) {
    if (o.inputs === undefined) {
        throw new Error("Expected inputs in Poseidon2HashAccumulate serialization");
    }
    ;
    return {
        inputs: o.inputs.map((v) => v),
    };
}
function toPoseidon2HashAccumulateResponse(o) {
    if (o.hash === undefined) {
        throw new Error("Expected hash in Poseidon2HashAccumulateResponse deserialization");
    }
    ;
    return {
        hash: o.hash,
    };
}
function fromPoseidon2HashAccumulateResponse(o) {
    if (o.hash === undefined) {
        throw new Error("Expected hash in Poseidon2HashAccumulateResponse serialization");
    }
    ;
    return {
        hash: o.hash,
    };
}
function toPoseidon2HashResponse(o) {
    if (o.hash === undefined) {
        throw new Error("Expected hash in Poseidon2HashResponse deserialization");
    }
    ;
    return {
        hash: o.hash,
    };
}
function fromPoseidon2HashResponse(o) {
    if (o.hash === undefined) {
        throw new Error("Expected hash in Poseidon2HashResponse serialization");
    }
    ;
    return {
        hash: o.hash,
    };
}
function toPoseidon2Permutation(o) {
    if (o.inputs === undefined) {
        throw new Error("Expected inputs in Poseidon2Permutation deserialization");
    }
    ;
    return {
        inputs: o.inputs.map((v) => v),
    };
}
function fromPoseidon2Permutation(o) {
    if (o.inputs === undefined) {
        throw new Error("Expected inputs in Poseidon2Permutation serialization");
    }
    ;
    return {
        inputs: o.inputs.map((v) => v),
    };
}
function toPoseidon2PermutationResponse(o) {
    if (o.outputs === undefined) {
        throw new Error("Expected outputs in Poseidon2PermutationResponse deserialization");
    }
    ;
    return {
        outputs: o.outputs.map((v) => v),
    };
}
function fromPoseidon2PermutationResponse(o) {
    if (o.outputs === undefined) {
        throw new Error("Expected outputs in Poseidon2PermutationResponse serialization");
    }
    ;
    return {
        outputs: o.outputs.map((v) => v),
    };
}
function toProofSystemSettings(o) {
    if (o.ipa_accumulation === undefined) {
        throw new Error("Expected ipa_accumulation in ProofSystemSettings deserialization");
    }
    if (o.oracle_hash_type === undefined) {
        throw new Error("Expected oracle_hash_type in ProofSystemSettings deserialization");
    }
    if (o.disable_zk === undefined) {
        throw new Error("Expected disable_zk in ProofSystemSettings deserialization");
    }
    if (o.optimized_solidity_verifier === undefined) {
        throw new Error("Expected optimized_solidity_verifier in ProofSystemSettings deserialization");
    }
    ;
    return {
        ipaAccumulation: o.ipa_accumulation,
        oracleHashType: o.oracle_hash_type,
        disableZk: o.disable_zk,
        optimizedSolidityVerifier: o.optimized_solidity_verifier,
    };
}
function fromProofSystemSettings(o) {
    if (o.ipaAccumulation === undefined) {
        throw new Error("Expected ipaAccumulation in ProofSystemSettings serialization");
    }
    if (o.oracleHashType === undefined) {
        throw new Error("Expected oracleHashType in ProofSystemSettings serialization");
    }
    if (o.disableZk === undefined) {
        throw new Error("Expected disableZk in ProofSystemSettings serialization");
    }
    if (o.optimizedSolidityVerifier === undefined) {
        throw new Error("Expected optimizedSolidityVerifier in ProofSystemSettings serialization");
    }
    ;
    return {
        ipa_accumulation: o.ipaAccumulation,
        oracle_hash_type: o.oracleHashType,
        disable_zk: o.disableZk,
        optimized_solidity_verifier: o.optimizedSolidityVerifier,
    };
}
function toSchnorrComputePublicKey(o) {
    if (o.private_key === undefined) {
        throw new Error("Expected private_key in SchnorrComputePublicKey deserialization");
    }
    ;
    return {
        privateKey: o.private_key,
    };
}
function fromSchnorrComputePublicKey(o) {
    if (o.privateKey === undefined) {
        throw new Error("Expected privateKey in SchnorrComputePublicKey serialization");
    }
    ;
    return {
        private_key: o.privateKey,
    };
}
function toSchnorrComputePublicKeyResponse(o) {
    if (o.public_key === undefined) {
        throw new Error("Expected public_key in SchnorrComputePublicKeyResponse deserialization");
    }
    ;
    return {
        publicKey: toAffineElement(o.public_key),
    };
}
function fromSchnorrComputePublicKeyResponse(o) {
    if (o.publicKey === undefined) {
        throw new Error("Expected publicKey in SchnorrComputePublicKeyResponse serialization");
    }
    ;
    return {
        public_key: fromAffineElement(o.publicKey),
    };
}
function toSchnorrConstructSignature(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in SchnorrConstructSignature deserialization");
    }
    if (o.private_key === undefined) {
        throw new Error("Expected private_key in SchnorrConstructSignature deserialization");
    }
    ;
    return {
        message: o.message,
        privateKey: o.private_key,
    };
}
function fromSchnorrConstructSignature(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in SchnorrConstructSignature serialization");
    }
    if (o.privateKey === undefined) {
        throw new Error("Expected privateKey in SchnorrConstructSignature serialization");
    }
    ;
    return {
        message: o.message,
        private_key: o.privateKey,
    };
}
function toSchnorrConstructSignatureResponse(o) {
    if (o.s === undefined) {
        throw new Error("Expected s in SchnorrConstructSignatureResponse deserialization");
    }
    if (o.e === undefined) {
        throw new Error("Expected e in SchnorrConstructSignatureResponse deserialization");
    }
    ;
    return {
        s: o.s,
        e: o.e,
    };
}
function fromSchnorrConstructSignatureResponse(o) {
    if (o.s === undefined) {
        throw new Error("Expected s in SchnorrConstructSignatureResponse serialization");
    }
    if (o.e === undefined) {
        throw new Error("Expected e in SchnorrConstructSignatureResponse serialization");
    }
    ;
    return {
        s: o.s,
        e: o.e,
    };
}
function toSchnorrVerifySignature(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in SchnorrVerifySignature deserialization");
    }
    if (o.public_key === undefined) {
        throw new Error("Expected public_key in SchnorrVerifySignature deserialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in SchnorrVerifySignature deserialization");
    }
    if (o.e === undefined) {
        throw new Error("Expected e in SchnorrVerifySignature deserialization");
    }
    ;
    return {
        message: o.message,
        publicKey: toAffineElement(o.public_key),
        s: o.s,
        e: o.e,
    };
}
function fromSchnorrVerifySignature(o) {
    if (o.message === undefined) {
        throw new Error("Expected message in SchnorrVerifySignature serialization");
    }
    if (o.publicKey === undefined) {
        throw new Error("Expected publicKey in SchnorrVerifySignature serialization");
    }
    if (o.s === undefined) {
        throw new Error("Expected s in SchnorrVerifySignature serialization");
    }
    if (o.e === undefined) {
        throw new Error("Expected e in SchnorrVerifySignature serialization");
    }
    ;
    return {
        message: o.message,
        public_key: fromAffineElement(o.publicKey),
        s: o.s,
        e: o.e,
    };
}
function toSchnorrVerifySignatureResponse(o) {
    if (o.verified === undefined) {
        throw new Error("Expected verified in SchnorrVerifySignatureResponse deserialization");
    }
    ;
    return {
        verified: o.verified,
    };
}
function fromSchnorrVerifySignatureResponse(o) {
    if (o.verified === undefined) {
        throw new Error("Expected verified in SchnorrVerifySignatureResponse serialization");
    }
    ;
    return {
        verified: o.verified,
    };
}
function toSecp256k1GetRandomFr(o) {
    if (o.dummy === undefined) {
        throw new Error("Expected dummy in Secp256k1GetRandomFr deserialization");
    }
    ;
    return {
        dummy: o.dummy,
    };
}
function fromSecp256k1GetRandomFr(o) {
    if (o.dummy === undefined) {
        throw new Error("Expected dummy in Secp256k1GetRandomFr serialization");
    }
    ;
    return {
        dummy: o.dummy,
    };
}
function toSecp256k1GetRandomFrResponse(o) {
    if (o.value === undefined) {
        throw new Error("Expected value in Secp256k1GetRandomFrResponse deserialization");
    }
    ;
    return {
        value: o.value,
    };
}
function fromSecp256k1GetRandomFrResponse(o) {
    if (o.value === undefined) {
        throw new Error("Expected value in Secp256k1GetRandomFrResponse serialization");
    }
    ;
    return {
        value: o.value,
    };
}
function toSecp256k1Mul(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in Secp256k1Mul deserialization");
    }
    if (o.scalar === undefined) {
        throw new Error("Expected scalar in Secp256k1Mul deserialization");
    }
    ;
    return {
        point: toAffineElement(o.point),
        scalar: o.scalar,
    };
}
function fromSecp256k1Mul(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in Secp256k1Mul serialization");
    }
    if (o.scalar === undefined) {
        throw new Error("Expected scalar in Secp256k1Mul serialization");
    }
    ;
    return {
        point: fromAffineElement(o.point),
        scalar: o.scalar,
    };
}
function toSecp256k1MulResponse(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in Secp256k1MulResponse deserialization");
    }
    ;
    return {
        point: toAffineElement(o.point),
    };
}
function fromSecp256k1MulResponse(o) {
    if (o.point === undefined) {
        throw new Error("Expected point in Secp256k1MulResponse serialization");
    }
    ;
    return {
        point: fromAffineElement(o.point),
    };
}
function toSecp256k1Reduce512(o) {
    if (o.input === undefined) {
        throw new Error("Expected input in Secp256k1Reduce512 deserialization");
    }
    ;
    return {
        input: o.input,
    };
}
function fromSecp256k1Reduce512(o) {
    if (o.input === undefined) {
        throw new Error("Expected input in Secp256k1Reduce512 serialization");
    }
    ;
    return {
        input: o.input,
    };
}
function toSecp256k1Reduce512Response(o) {
    if (o.value === undefined) {
        throw new Error("Expected value in Secp256k1Reduce512Response deserialization");
    }
    ;
    return {
        value: o.value,
    };
}
function fromSecp256k1Reduce512Response(o) {
    if (o.value === undefined) {
        throw new Error("Expected value in Secp256k1Reduce512Response serialization");
    }
    ;
    return {
        value: o.value,
    };
}
function toShutdown(o) {
    return {};
}
function fromShutdown(o) {
    return {};
}
function toShutdownResponse(o) {
    return {};
}
function fromShutdownResponse(o) {
    return {};
}
function toSrsInitGrumpkinSrs(o) {
    if (o.points_buf === undefined) {
        throw new Error("Expected points_buf in SrsInitGrumpkinSrs deserialization");
    }
    if (o.num_points === undefined) {
        throw new Error("Expected num_points in SrsInitGrumpkinSrs deserialization");
    }
    ;
    return {
        pointsBuf: o.points_buf,
        numPoints: o.num_points,
    };
}
function fromSrsInitGrumpkinSrs(o) {
    if (o.pointsBuf === undefined) {
        throw new Error("Expected pointsBuf in SrsInitGrumpkinSrs serialization");
    }
    if (o.numPoints === undefined) {
        throw new Error("Expected numPoints in SrsInitGrumpkinSrs serialization");
    }
    ;
    return {
        points_buf: o.pointsBuf,
        num_points: o.numPoints,
    };
}
function toSrsInitGrumpkinSrsResponse(o) {
    if (o.dummy === undefined) {
        throw new Error("Expected dummy in SrsInitGrumpkinSrsResponse deserialization");
    }
    ;
    return {
        dummy: o.dummy,
    };
}
function fromSrsInitGrumpkinSrsResponse(o) {
    if (o.dummy === undefined) {
        throw new Error("Expected dummy in SrsInitGrumpkinSrsResponse serialization");
    }
    ;
    return {
        dummy: o.dummy,
    };
}
function toSrsInitSrs(o) {
    if (o.points_buf === undefined) {
        throw new Error("Expected points_buf in SrsInitSrs deserialization");
    }
    if (o.num_points === undefined) {
        throw new Error("Expected num_points in SrsInitSrs deserialization");
    }
    if (o.g2_point === undefined) {
        throw new Error("Expected g2_point in SrsInitSrs deserialization");
    }
    ;
    return {
        pointsBuf: o.points_buf,
        numPoints: o.num_points,
        g2Point: o.g2_point,
    };
}
function fromSrsInitSrs(o) {
    if (o.pointsBuf === undefined) {
        throw new Error("Expected pointsBuf in SrsInitSrs serialization");
    }
    if (o.numPoints === undefined) {
        throw new Error("Expected numPoints in SrsInitSrs serialization");
    }
    if (o.g2Point === undefined) {
        throw new Error("Expected g2Point in SrsInitSrs serialization");
    }
    ;
    return {
        points_buf: o.pointsBuf,
        num_points: o.numPoints,
        g2_point: o.g2Point,
    };
}
function toSrsInitSrsResponse(o) {
    if (o.dummy === undefined) {
        throw new Error("Expected dummy in SrsInitSrsResponse deserialization");
    }
    ;
    return {
        dummy: o.dummy,
    };
}
function fromSrsInitSrsResponse(o) {
    if (o.dummy === undefined) {
        throw new Error("Expected dummy in SrsInitSrsResponse serialization");
    }
    ;
    return {
        dummy: o.dummy,
    };
}
function toVkAsFields(o) {
    if (o.verification_key === undefined) {
        throw new Error("Expected verification_key in VkAsFields deserialization");
    }
    ;
    return {
        verificationKey: o.verification_key,
    };
}
function fromVkAsFields(o) {
    if (o.verificationKey === undefined) {
        throw new Error("Expected verificationKey in VkAsFields serialization");
    }
    ;
    return {
        verification_key: o.verificationKey,
    };
}
function toVkAsFieldsResponse(o) {
    if (o.fields === undefined) {
        throw new Error("Expected fields in VkAsFieldsResponse deserialization");
    }
    ;
    return {
        fields: o.fields.map((v) => v),
    };
}
function fromVkAsFieldsResponse(o) {
    if (o.fields === undefined) {
        throw new Error("Expected fields in VkAsFieldsResponse serialization");
    }
    ;
    return {
        fields: o.fields.map((v) => v),
    };
}
