// @aztec/bb.js/dest/node-cjs/cbind/generated/sync.js
"use strict";
// AUTOGENERATED FILE - DO NOT EDIT
Object.defineProperty(exports, "__esModule", { value: true });
exports.SyncApi = void 0;
const msgpackr_1 = require("msgpackr");
const api_types_js_1 = require("./api_types.js");
function msgpackCall(backend, input) { const inputBuffer = new msgpackr_1.Encoder({ useRecords: false }).pack(input); const encodedResult = backend.call(inputBuffer); return new msgpackr_1.Decoder({ useRecords: false }).unpack(encodedResult); }
class SyncApi {
    constructor(backend) {
        this.backend = backend;
    }
    circuitProve(command) {
        const msgpackCommand = (0, api_types_js_1.fromCircuitProve)(command);
        const [variantName, result] = msgpackCall(this.backend, [["CircuitProve", msgpackCommand]]);
        if (variantName !== 'CircuitProveResponse') {
            throw new Error(`Expected variant name 'CircuitProveResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toCircuitProveResponse)(result);
    }
    circuitComputeVk(command) {
        const msgpackCommand = (0, api_types_js_1.fromCircuitComputeVk)(command);
        const [variantName, result] = msgpackCall(this.backend, [["CircuitComputeVk", msgpackCommand]]);
        if (variantName !== 'CircuitComputeVkResponse') {
            throw new Error(`Expected variant name 'CircuitComputeVkResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toCircuitComputeVkResponse)(result);
    }
    circuitStats(command) {
        const msgpackCommand = (0, api_types_js_1.fromCircuitStats)(command);
        const [variantName, result] = msgpackCall(this.backend, [["CircuitStats", msgpackCommand]]);
        if (variantName !== 'CircuitInfoResponse') {
            throw new Error(`Expected variant name 'CircuitInfoResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toCircuitInfoResponse)(result);
    }
    circuitVerify(command) {
        const msgpackCommand = (0, api_types_js_1.fromCircuitVerify)(command);
        const [variantName, result] = msgpackCall(this.backend, [["CircuitVerify", msgpackCommand]]);
        if (variantName !== 'CircuitVerifyResponse') {
            throw new Error(`Expected variant name 'CircuitVerifyResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toCircuitVerifyResponse)(result);
    }
    clientIvcComputeStandaloneVk(command) {
        const msgpackCommand = (0, api_types_js_1.fromClientIvcComputeStandaloneVk)(command);
        const [variantName, result] = msgpackCall(this.backend, [["ClientIvcComputeStandaloneVk", msgpackCommand]]);
        if (variantName !== 'ClientIvcComputeStandaloneVkResponse') {
            throw new Error(`Expected variant name 'ClientIvcComputeStandaloneVkResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toClientIvcComputeStandaloneVkResponse)(result);
    }
    clientIvcComputeIvcVk(command) {
        const msgpackCommand = (0, api_types_js_1.fromClientIvcComputeIvcVk)(command);
        const [variantName, result] = msgpackCall(this.backend, [["ClientIvcComputeIvcVk", msgpackCommand]]);
        if (variantName !== 'ClientIvcComputeIvcVkResponse') {
            throw new Error(`Expected variant name 'ClientIvcComputeIvcVkResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toClientIvcComputeIvcVkResponse)(result);
    }
    clientIvcStart(command) {
        const msgpackCommand = (0, api_types_js_1.fromClientIvcStart)(command);
        const [variantName, result] = msgpackCall(this.backend, [["ClientIvcStart", msgpackCommand]]);
        if (variantName !== 'ClientIvcStartResponse') {
            throw new Error(`Expected variant name 'ClientIvcStartResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toClientIvcStartResponse)(result);
    }
    clientIvcLoad(command) {
        const msgpackCommand = (0, api_types_js_1.fromClientIvcLoad)(command);
        const [variantName, result] = msgpackCall(this.backend, [["ClientIvcLoad", msgpackCommand]]);
        if (variantName !== 'ClientIvcLoadResponse') {
            throw new Error(`Expected variant name 'ClientIvcLoadResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toClientIvcLoadResponse)(result);
    }
    clientIvcAccumulate(command) {
        const msgpackCommand = (0, api_types_js_1.fromClientIvcAccumulate)(command);
        const [variantName, result] = msgpackCall(this.backend, [["ClientIvcAccumulate", msgpackCommand]]);
        if (variantName !== 'ClientIvcAccumulateResponse') {
            throw new Error(`Expected variant name 'ClientIvcAccumulateResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toClientIvcAccumulateResponse)(result);
    }
    clientIvcProve(command) {
        const msgpackCommand = (0, api_types_js_1.fromClientIvcProve)(command);
        const [variantName, result] = msgpackCall(this.backend, [["ClientIvcProve", msgpackCommand]]);
        if (variantName !== 'ClientIvcProveResponse') {
            throw new Error(`Expected variant name 'ClientIvcProveResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toClientIvcProveResponse)(result);
    }
    clientIvcVerify(command) {
        const msgpackCommand = (0, api_types_js_1.fromClientIvcVerify)(command);
        const [variantName, result] = msgpackCall(this.backend, [["ClientIvcVerify", msgpackCommand]]);
        if (variantName !== 'ClientIvcVerifyResponse') {
            throw new Error(`Expected variant name 'ClientIvcVerifyResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toClientIvcVerifyResponse)(result);
    }
    vkAsFields(command) {
        const msgpackCommand = (0, api_types_js_1.fromVkAsFields)(command);
        const [variantName, result] = msgpackCall(this.backend, [["VkAsFields", msgpackCommand]]);
        if (variantName !== 'VkAsFieldsResponse') {
            throw new Error(`Expected variant name 'VkAsFieldsResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toVkAsFieldsResponse)(result);
    }
    megaVkAsFields(command) {
        const msgpackCommand = (0, api_types_js_1.fromMegaVkAsFields)(command);
        const [variantName, result] = msgpackCall(this.backend, [["MegaVkAsFields", msgpackCommand]]);
        if (variantName !== 'MegaVkAsFieldsResponse') {
            throw new Error(`Expected variant name 'MegaVkAsFieldsResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toMegaVkAsFieldsResponse)(result);
    }
    circuitWriteSolidityVerifier(command) {
        const msgpackCommand = (0, api_types_js_1.fromCircuitWriteSolidityVerifier)(command);
        const [variantName, result] = msgpackCall(this.backend, [["CircuitWriteSolidityVerifier", msgpackCommand]]);
        if (variantName !== 'CircuitWriteSolidityVerifierResponse') {
            throw new Error(`Expected variant name 'CircuitWriteSolidityVerifierResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toCircuitWriteSolidityVerifierResponse)(result);
    }
    clientIvcCheckPrecomputedVk(command) {
        const msgpackCommand = (0, api_types_js_1.fromClientIvcCheckPrecomputedVk)(command);
        const [variantName, result] = msgpackCall(this.backend, [["ClientIvcCheckPrecomputedVk", msgpackCommand]]);
        if (variantName !== 'ClientIvcCheckPrecomputedVkResponse') {
            throw new Error(`Expected variant name 'ClientIvcCheckPrecomputedVkResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toClientIvcCheckPrecomputedVkResponse)(result);
    }
    clientIvcStats(command) {
        const msgpackCommand = (0, api_types_js_1.fromClientIvcStats)(command);
        const [variantName, result] = msgpackCall(this.backend, [["ClientIvcStats", msgpackCommand]]);
        if (variantName !== 'ClientIvcStatsResponse') {
            throw new Error(`Expected variant name 'ClientIvcStatsResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toClientIvcStatsResponse)(result);
    }
    poseidon2Hash(command) {
        const msgpackCommand = (0, api_types_js_1.fromPoseidon2Hash)(command);
        const [variantName, result] = msgpackCall(this.backend, [["Poseidon2Hash", msgpackCommand]]);
        if (variantName !== 'Poseidon2HashResponse') {
            throw new Error(`Expected variant name 'Poseidon2HashResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toPoseidon2HashResponse)(result);
    }
    poseidon2Permutation(command) {
        const msgpackCommand = (0, api_types_js_1.fromPoseidon2Permutation)(command);
        const [variantName, result] = msgpackCall(this.backend, [["Poseidon2Permutation", msgpackCommand]]);
        if (variantName !== 'Poseidon2PermutationResponse') {
            throw new Error(`Expected variant name 'Poseidon2PermutationResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toPoseidon2PermutationResponse)(result);
    }
    poseidon2HashAccumulate(command) {
        const msgpackCommand = (0, api_types_js_1.fromPoseidon2HashAccumulate)(command);
        const [variantName, result] = msgpackCall(this.backend, [["Poseidon2HashAccumulate", msgpackCommand]]);
        if (variantName !== 'Poseidon2HashAccumulateResponse') {
            throw new Error(`Expected variant name 'Poseidon2HashAccumulateResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toPoseidon2HashAccumulateResponse)(result);
    }
    pedersenCommit(command) {
        const msgpackCommand = (0, api_types_js_1.fromPedersenCommit)(command);
        const [variantName, result] = msgpackCall(this.backend, [["PedersenCommit", msgpackCommand]]);
        if (variantName !== 'PedersenCommitResponse') {
            throw new Error(`Expected variant name 'PedersenCommitResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toPedersenCommitResponse)(result);
    }
    pedersenHash(command) {
        const msgpackCommand = (0, api_types_js_1.fromPedersenHash)(command);
        const [variantName, result] = msgpackCall(this.backend, [["PedersenHash", msgpackCommand]]);
        if (variantName !== 'PedersenHashResponse') {
            throw new Error(`Expected variant name 'PedersenHashResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toPedersenHashResponse)(result);
    }
    pedersenHashBuffer(command) {
        const msgpackCommand = (0, api_types_js_1.fromPedersenHashBuffer)(command);
        const [variantName, result] = msgpackCall(this.backend, [["PedersenHashBuffer", msgpackCommand]]);
        if (variantName !== 'PedersenHashBufferResponse') {
            throw new Error(`Expected variant name 'PedersenHashBufferResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toPedersenHashBufferResponse)(result);
    }
    blake2s(command) {
        const msgpackCommand = (0, api_types_js_1.fromBlake2s)(command);
        const [variantName, result] = msgpackCall(this.backend, [["Blake2s", msgpackCommand]]);
        if (variantName !== 'Blake2sResponse') {
            throw new Error(`Expected variant name 'Blake2sResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toBlake2sResponse)(result);
    }
    blake2sToField(command) {
        const msgpackCommand = (0, api_types_js_1.fromBlake2sToField)(command);
        const [variantName, result] = msgpackCall(this.backend, [["Blake2sToField", msgpackCommand]]);
        if (variantName !== 'Blake2sToFieldResponse') {
            throw new Error(`Expected variant name 'Blake2sToFieldResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toBlake2sToFieldResponse)(result);
    }
    aesEncrypt(command) {
        const msgpackCommand = (0, api_types_js_1.fromAesEncrypt)(command);
        const [variantName, result] = msgpackCall(this.backend, [["AesEncrypt", msgpackCommand]]);
        if (variantName !== 'AesEncryptResponse') {
            throw new Error(`Expected variant name 'AesEncryptResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toAesEncryptResponse)(result);
    }
    aesDecrypt(command) {
        const msgpackCommand = (0, api_types_js_1.fromAesDecrypt)(command);
        const [variantName, result] = msgpackCall(this.backend, [["AesDecrypt", msgpackCommand]]);
        if (variantName !== 'AesDecryptResponse') {
            throw new Error(`Expected variant name 'AesDecryptResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toAesDecryptResponse)(result);
    }
    grumpkinMul(command) {
        const msgpackCommand = (0, api_types_js_1.fromGrumpkinMul)(command);
        const [variantName, result] = msgpackCall(this.backend, [["GrumpkinMul", msgpackCommand]]);
        if (variantName !== 'GrumpkinMulResponse') {
            throw new Error(`Expected variant name 'GrumpkinMulResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toGrumpkinMulResponse)(result);
    }
    grumpkinAdd(command) {
        const msgpackCommand = (0, api_types_js_1.fromGrumpkinAdd)(command);
        const [variantName, result] = msgpackCall(this.backend, [["GrumpkinAdd", msgpackCommand]]);
        if (variantName !== 'GrumpkinAddResponse') {
            throw new Error(`Expected variant name 'GrumpkinAddResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toGrumpkinAddResponse)(result);
    }
    grumpkinBatchMul(command) {
        const msgpackCommand = (0, api_types_js_1.fromGrumpkinBatchMul)(command);
        const [variantName, result] = msgpackCall(this.backend, [["GrumpkinBatchMul", msgpackCommand]]);
        if (variantName !== 'GrumpkinBatchMulResponse') {
            throw new Error(`Expected variant name 'GrumpkinBatchMulResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toGrumpkinBatchMulResponse)(result);
    }
    grumpkinGetRandomFr(command) {
        const msgpackCommand = (0, api_types_js_1.fromGrumpkinGetRandomFr)(command);
        const [variantName, result] = msgpackCall(this.backend, [["GrumpkinGetRandomFr", msgpackCommand]]);
        if (variantName !== 'GrumpkinGetRandomFrResponse') {
            throw new Error(`Expected variant name 'GrumpkinGetRandomFrResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toGrumpkinGetRandomFrResponse)(result);
    }
    grumpkinReduce512(command) {
        const msgpackCommand = (0, api_types_js_1.fromGrumpkinReduce512)(command);
        const [variantName, result] = msgpackCall(this.backend, [["GrumpkinReduce512", msgpackCommand]]);
        if (variantName !== 'GrumpkinReduce512Response') {
            throw new Error(`Expected variant name 'GrumpkinReduce512Response' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toGrumpkinReduce512Response)(result);
    }
    secp256k1Mul(command) {
        const msgpackCommand = (0, api_types_js_1.fromSecp256k1Mul)(command);
        const [variantName, result] = msgpackCall(this.backend, [["Secp256k1Mul", msgpackCommand]]);
        if (variantName !== 'Secp256k1MulResponse') {
            throw new Error(`Expected variant name 'Secp256k1MulResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toSecp256k1MulResponse)(result);
    }
    secp256k1GetRandomFr(command) {
        const msgpackCommand = (0, api_types_js_1.fromSecp256k1GetRandomFr)(command);
        const [variantName, result] = msgpackCall(this.backend, [["Secp256k1GetRandomFr", msgpackCommand]]);
        if (variantName !== 'Secp256k1GetRandomFrResponse') {
            throw new Error(`Expected variant name 'Secp256k1GetRandomFrResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toSecp256k1GetRandomFrResponse)(result);
    }
    secp256k1Reduce512(command) {
        const msgpackCommand = (0, api_types_js_1.fromSecp256k1Reduce512)(command);
        const [variantName, result] = msgpackCall(this.backend, [["Secp256k1Reduce512", msgpackCommand]]);
        if (variantName !== 'Secp256k1Reduce512Response') {
            throw new Error(`Expected variant name 'Secp256k1Reduce512Response' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toSecp256k1Reduce512Response)(result);
    }
    bn254FrSqrt(command) {
        const msgpackCommand = (0, api_types_js_1.fromBn254FrSqrt)(command);
        const [variantName, result] = msgpackCall(this.backend, [["Bn254FrSqrt", msgpackCommand]]);
        if (variantName !== 'Bn254FrSqrtResponse') {
            throw new Error(`Expected variant name 'Bn254FrSqrtResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toBn254FrSqrtResponse)(result);
    }
    schnorrComputePublicKey(command) {
        const msgpackCommand = (0, api_types_js_1.fromSchnorrComputePublicKey)(command);
        const [variantName, result] = msgpackCall(this.backend, [["SchnorrComputePublicKey", msgpackCommand]]);
        if (variantName !== 'SchnorrComputePublicKeyResponse') {
            throw new Error(`Expected variant name 'SchnorrComputePublicKeyResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toSchnorrComputePublicKeyResponse)(result);
    }
    schnorrConstructSignature(command) {
        const msgpackCommand = (0, api_types_js_1.fromSchnorrConstructSignature)(command);
        const [variantName, result] = msgpackCall(this.backend, [["SchnorrConstructSignature", msgpackCommand]]);
        if (variantName !== 'SchnorrConstructSignatureResponse') {
            throw new Error(`Expected variant name 'SchnorrConstructSignatureResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toSchnorrConstructSignatureResponse)(result);
    }
    schnorrVerifySignature(command) {
        const msgpackCommand = (0, api_types_js_1.fromSchnorrVerifySignature)(command);
        const [variantName, result] = msgpackCall(this.backend, [["SchnorrVerifySignature", msgpackCommand]]);
        if (variantName !== 'SchnorrVerifySignatureResponse') {
            throw new Error(`Expected variant name 'SchnorrVerifySignatureResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toSchnorrVerifySignatureResponse)(result);
    }
    ecdsaSecp256k1ComputePublicKey(command) {
        const msgpackCommand = (0, api_types_js_1.fromEcdsaSecp256k1ComputePublicKey)(command);
        const [variantName, result] = msgpackCall(this.backend, [["EcdsaSecp256k1ComputePublicKey", msgpackCommand]]);
        if (variantName !== 'EcdsaSecp256k1ComputePublicKeyResponse') {
            throw new Error(`Expected variant name 'EcdsaSecp256k1ComputePublicKeyResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toEcdsaSecp256k1ComputePublicKeyResponse)(result);
    }
    ecdsaSecp256r1ComputePublicKey(command) {
        const msgpackCommand = (0, api_types_js_1.fromEcdsaSecp256r1ComputePublicKey)(command);
        const [variantName, result] = msgpackCall(this.backend, [["EcdsaSecp256r1ComputePublicKey", msgpackCommand]]);
        if (variantName !== 'EcdsaSecp256r1ComputePublicKeyResponse') {
            throw new Error(`Expected variant name 'EcdsaSecp256r1ComputePublicKeyResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toEcdsaSecp256r1ComputePublicKeyResponse)(result);
    }
    ecdsaSecp256k1ConstructSignature(command) {
        const msgpackCommand = (0, api_types_js_1.fromEcdsaSecp256k1ConstructSignature)(command);
        const [variantName, result] = msgpackCall(this.backend, [["EcdsaSecp256k1ConstructSignature", msgpackCommand]]);
        if (variantName !== 'EcdsaSecp256k1ConstructSignatureResponse') {
            throw new Error(`Expected variant name 'EcdsaSecp256k1ConstructSignatureResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toEcdsaSecp256k1ConstructSignatureResponse)(result);
    }
    ecdsaSecp256r1ConstructSignature(command) {
        const msgpackCommand = (0, api_types_js_1.fromEcdsaSecp256r1ConstructSignature)(command);
        const [variantName, result] = msgpackCall(this.backend, [["EcdsaSecp256r1ConstructSignature", msgpackCommand]]);
        if (variantName !== 'EcdsaSecp256r1ConstructSignatureResponse') {
            throw new Error(`Expected variant name 'EcdsaSecp256r1ConstructSignatureResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toEcdsaSecp256r1ConstructSignatureResponse)(result);
    }
    ecdsaSecp256k1RecoverPublicKey(command) {
        const msgpackCommand = (0, api_types_js_1.fromEcdsaSecp256k1RecoverPublicKey)(command);
        const [variantName, result] = msgpackCall(this.backend, [["EcdsaSecp256k1RecoverPublicKey", msgpackCommand]]);
        if (variantName !== 'EcdsaSecp256k1RecoverPublicKeyResponse') {
            throw new Error(`Expected variant name 'EcdsaSecp256k1RecoverPublicKeyResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toEcdsaSecp256k1RecoverPublicKeyResponse)(result);
    }
    ecdsaSecp256r1RecoverPublicKey(command) {
        const msgpackCommand = (0, api_types_js_1.fromEcdsaSecp256r1RecoverPublicKey)(command);
        const [variantName, result] = msgpackCall(this.backend, [["EcdsaSecp256r1RecoverPublicKey", msgpackCommand]]);
        if (variantName !== 'EcdsaSecp256r1RecoverPublicKeyResponse') {
            throw new Error(`Expected variant name 'EcdsaSecp256r1RecoverPublicKeyResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toEcdsaSecp256r1RecoverPublicKeyResponse)(result);
    }
    ecdsaSecp256k1VerifySignature(command) {
        const msgpackCommand = (0, api_types_js_1.fromEcdsaSecp256k1VerifySignature)(command);
        const [variantName, result] = msgpackCall(this.backend, [["EcdsaSecp256k1VerifySignature", msgpackCommand]]);
        if (variantName !== 'EcdsaSecp256k1VerifySignatureResponse') {
            throw new Error(`Expected variant name 'EcdsaSecp256k1VerifySignatureResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toEcdsaSecp256k1VerifySignatureResponse)(result);
    }
    ecdsaSecp256r1VerifySignature(command) {
        const msgpackCommand = (0, api_types_js_1.fromEcdsaSecp256r1VerifySignature)(command);
        const [variantName, result] = msgpackCall(this.backend, [["EcdsaSecp256r1VerifySignature", msgpackCommand]]);
        if (variantName !== 'EcdsaSecp256r1VerifySignatureResponse') {
            throw new Error(`Expected variant name 'EcdsaSecp256r1VerifySignatureResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toEcdsaSecp256r1VerifySignatureResponse)(result);
    }
    srsInitSrs(command) {
        const msgpackCommand = (0, api_types_js_1.fromSrsInitSrs)(command);
        const [variantName, result] = msgpackCall(this.backend, [["SrsInitSrs", msgpackCommand]]);
        if (variantName !== 'SrsInitSrsResponse') {
            throw new Error(`Expected variant name 'SrsInitSrsResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toSrsInitSrsResponse)(result);
    }
    srsInitGrumpkinSrs(command) {
        const msgpackCommand = (0, api_types_js_1.fromSrsInitGrumpkinSrs)(command);
        const [variantName, result] = msgpackCall(this.backend, [["SrsInitGrumpkinSrs", msgpackCommand]]);
        if (variantName !== 'SrsInitGrumpkinSrsResponse') {
            throw new Error(`Expected variant name 'SrsInitGrumpkinSrsResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toSrsInitGrumpkinSrsResponse)(result);
    }
    shutdown(command) {
        const msgpackCommand = (0, api_types_js_1.fromShutdown)(command);
        const [variantName, result] = msgpackCall(this.backend, [["Shutdown", msgpackCommand]]);
        if (variantName !== 'ShutdownResponse') {
            throw new Error(`Expected variant name 'ShutdownResponse' but got '${variantName}'`);
        }
        return (0, api_types_js_1.toShutdownResponse)(result);
    }
    destroy() {
        if (this.backend.destroy)
            this.backend.destroy();
    }
}
exports.SyncApi = SyncApi;
