// @aztec/stdlib/dest/p2p/topic_type.js
import { P2PClientType } from './client_type.js';
/** Create Topic String
 *
 * The topic channel identifier
 * @param topicType
 * @returns
 */ export function createTopicString(topicType, protocolVersion) {
    return `/aztec/${TopicType[topicType]}/${protocolVersion}`;
}
/**
 *
 */ export var TopicType = /*#__PURE__*/ function(TopicType) {
    TopicType["tx"] = "tx";
    TopicType["block_proposal"] = "block_proposal";
    TopicType["block_attestation"] = "block_attestation";
    return TopicType;
}({});
export function getTopicTypeForClientType(clientType) {
    if (clientType === P2PClientType.Full) {
        return Object.values(TopicType);
    } else if (clientType === P2PClientType.Prover) {
        return [
            "tx",
            "block_proposal"
        ];
    } else {
        const _ = clientType;
        return [
            "tx"
        ];
    }
}
export function getTopicsForClientAndConfig(clientType, disableTransactions) {
    const topics = getTopicTypeForClientType(clientType);
    if (disableTransactions) {
        return topics.filter((topic)=>topic !== "tx");
    }
    return topics;
}
/**
 * Convert the topic string into a set of labels
 *
 * In the form:
 * {
 *  "/aztec/tx/0.1.0": "tx",
 *  ...
 * }
 */ export function metricsTopicStrToLabels(protocolVersion) {
    const topicStrToLabel = new Map();
    for(const topic in TopicType){
        topicStrToLabel.set(createTopicString(TopicType[topic], protocolVersion), topic);
    }
    return topicStrToLabel;
}
