// @aztec/stdlib/dest/rollup/checkpoint_constant_data.js
import { EthAddress } from '@aztec/foundation/eth-address';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { AztecAddress } from '../aztec-address/index.js';
import { GasFees } from '../gas/gas_fees.js';
/**
 * Constants that are the same for the entire checkpoint.
 */ export class CheckpointConstantData {
    chainId;
    version;
    vkTreeRoot;
    protocolContractsHash;
    proverId;
    slotNumber;
    coinbase;
    feeRecipient;
    gasFees;
    constructor(/** ChainId of the rollup. */ chainId, /** Version of the rollup. */ version, /** Root of the verification key tree. */ vkTreeRoot, /** Hash of the protocol contracts list. */ protocolContractsHash, /** Identifier of the prover. */ proverId, /** Slot number of the checkpoint. */ slotNumber, /** Coinbase address of the rollup. */ coinbase, /** Address to receive fees. */ feeRecipient, /** Global gas fees for this checkpoint. */ gasFees){
        this.chainId = chainId;
        this.version = version;
        this.vkTreeRoot = vkTreeRoot;
        this.protocolContractsHash = protocolContractsHash;
        this.proverId = proverId;
        this.slotNumber = slotNumber;
        this.coinbase = coinbase;
        this.feeRecipient = feeRecipient;
        this.gasFees = gasFees;
    }
    static from(fields) {
        return new CheckpointConstantData(...CheckpointConstantData.getFields(fields));
    }
    static getFields(fields) {
        return [
            fields.chainId,
            fields.version,
            fields.vkTreeRoot,
            fields.protocolContractsHash,
            fields.proverId,
            fields.slotNumber,
            fields.coinbase,
            fields.feeRecipient,
            fields.gasFees
        ];
    }
    static empty() {
        return new CheckpointConstantData(Fr.ZERO, Fr.ZERO, Fr.ZERO, Fr.ZERO, Fr.ZERO, Fr.ZERO, EthAddress.ZERO, AztecAddress.ZERO, GasFees.empty());
    }
    toBuffer() {
        return serializeToBuffer(...CheckpointConstantData.getFields(this));
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new CheckpointConstantData(Fr.fromBuffer(reader), Fr.fromBuffer(reader), Fr.fromBuffer(reader), Fr.fromBuffer(reader), Fr.fromBuffer(reader), Fr.fromBuffer(reader), reader.readObject(EthAddress), reader.readObject(AztecAddress), reader.readObject(GasFees));
    }
}
