// @aztec/stdlib/dest/rollup/public_tube_public_inputs.js
import { Fr } from '@aztec/foundation/fields';
import { bufferSchemaFor } from '@aztec/foundation/schemas';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { bufferToHex, hexToBuffer } from '@aztec/foundation/string';
import { PrivateToPublicKernelCircuitPublicInputs } from '../kernel/private_to_public_kernel_circuit_public_inputs.js';
export class PublicTubePublicInputs {
    privateTail;
    proverId;
    constructor(privateTail, proverId){
        this.privateTail = privateTail;
        this.proverId = proverId;
    }
    static from(fields) {
        return new PublicTubePublicInputs(...PublicTubePublicInputs.getFields(fields));
    }
    static getFields(fields) {
        return [
            fields.privateTail,
            fields.proverId
        ];
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PublicTubePublicInputs(reader.readObject(PrivateToPublicKernelCircuitPublicInputs), reader.readObject(Fr));
    }
    toBuffer() {
        return serializeToBuffer(...PublicTubePublicInputs.getFields(this));
    }
    static fromString(str) {
        return PublicTubePublicInputs.fromBuffer(hexToBuffer(str));
    }
    toString() {
        return bufferToHex(this.toBuffer());
    }
    /** Returns a representation for JSON serialization. */ toJSON() {
        return this.toBuffer();
    }
    /** Creates an instance from a string. */ static get schema() {
        return bufferSchemaFor(PublicTubePublicInputs);
    }
}
